// ngx-editor/esm2022/lib/modules/menu/menu.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { MenuComponent } from './menu.component';
import { ToggleCommandComponent } from './toggle-command/toggle-command.component';
import { LinkComponent } from './link/link.component';
import { DropdownComponent } from './dropdown/dropdown.component';
import { ImageComponent } from './image/image.component';
import { ColorPickerComponent } from './color-picker/color-picker.component';
import { FloatingMenuComponent } from './floating-menu/floating-menu.component';
import { BubbleComponent } from './bubble/bubble.component';
import { SanitizeHtmlPipe } from '../../pipes/sanitize/sanitize-html.pipe';
import { InsertCommandComponent } from './insert-command/insert-command.component';
import * as i0 from "@angular/core";
export class MenuModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.8", ngImport: i0, type: MenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.8", ngImport: i0, type: MenuModule, declarations: [
            // pipes
            SanitizeHtmlPipe,
            // components
            MenuComponent,
            ToggleCommandComponent,
            InsertCommandComponent,
            LinkComponent,
            DropdownComponent,
            ImageComponent,
            ColorPickerComponent,
            FloatingMenuComponent,
            BubbleComponent], imports: [CommonModule,
            ReactiveFormsModule], exports: [MenuComponent,
            FloatingMenuComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.8", ngImport: i0, type: MenuModule, providers: [
            SanitizeHtmlPipe,
        ], imports: [CommonModule,
            ReactiveFormsModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.8", ngImport: i0, type: MenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                    ],
                    declarations: [
                        // pipes
                        SanitizeHtmlPipe,
                        // components
                        MenuComponent,
                        ToggleCommandComponent,
                        InsertCommandComponent,
                        LinkComponent,
                        DropdownComponent,
                        ImageComponent,
                        ColorPickerComponent,
                        FloatingMenuComponent,
                        BubbleComponent,
                    ],
                    providers: [
                        SanitizeHtmlPipe,
                    ],
                    exports: [
                        MenuComponent,
                        FloatingMenuComponent,
                    ],
                }]
        }] });
