// devextreme-angular/fesm2015/devextreme-angular.js
import { DxTemplateModule } from 'devextreme-angular/core';
export * from 'devextreme-angular/core';
import * as i0 from '@angular/core';
import { NgModule } from '@angular/core';
import { DxAccordionModule } from 'devextreme-angular/ui/accordion';
export { DxAccordionComponent, DxAccordionModule } from 'devextreme-angular/ui/accordion';
import { DxActionSheetModule } from 'devextreme-angular/ui/action-sheet';
export { DxActionSheetComponent, DxActionSheetModule } from 'devextreme-angular/ui/action-sheet';
import { DxAutocompleteModule } from 'devextreme-angular/ui/autocomplete';
export { DxAutocompleteComponent, DxAutocompleteModule } from 'devextreme-angular/ui/autocomplete';
import { DxBarGaugeModule } from 'devextreme-angular/ui/bar-gauge';
export { DxBarGaugeComponent, DxBarGaugeModule } from 'devextreme-angular/ui/bar-gauge';
import { DxBoxModule } from 'devextreme-angular/ui/box';
export { DxBoxComponent, DxBoxModule } from 'devextreme-angular/ui/box';
import { DxBulletModule } from 'devextreme-angular/ui/bullet';
export { DxBulletComponent, DxBulletModule } from 'devextreme-angular/ui/bullet';
import { DxButtonModule } from 'devextreme-angular/ui/button';
export { DxButtonComponent, DxButtonModule } from 'devextreme-angular/ui/button';
import { DxButtonGroupModule } from 'devextreme-angular/ui/button-group';
export { DxButtonGroupComponent, DxButtonGroupModule } from 'devextreme-angular/ui/button-group';
import { DxCalendarModule } from 'devextreme-angular/ui/calendar';
export { DxCalendarComponent, DxCalendarModule } from 'devextreme-angular/ui/calendar';
import { DxChartModule } from 'devextreme-angular/ui/chart';
export { DxChartComponent, DxChartModule } from 'devextreme-angular/ui/chart';
import { DxCheckBoxModule } from 'devextreme-angular/ui/check-box';
export { DxCheckBoxComponent, DxCheckBoxModule } from 'devextreme-angular/ui/check-box';
import { DxCircularGaugeModule } from 'devextreme-angular/ui/circular-gauge';
export { DxCircularGaugeComponent, DxCircularGaugeModule } from 'devextreme-angular/ui/circular-gauge';
import { DxColorBoxModule } from 'devextreme-angular/ui/color-box';
export { DxColorBoxComponent, DxColorBoxModule } from 'devextreme-angular/ui/color-box';
import { DxContextMenuModule } from 'devextreme-angular/ui/context-menu';
export { DxContextMenuComponent, DxContextMenuModule } from 'devextreme-angular/ui/context-menu';
import { DxDataGridModule } from 'devextreme-angular/ui/data-grid';
export { DxDataGridComponent, DxDataGridModule } from 'devextreme-angular/ui/data-grid';
import { DxDateBoxModule } from 'devextreme-angular/ui/date-box';
export { DxDateBoxComponent, DxDateBoxModule } from 'devextreme-angular/ui/date-box';
import { DxDateRangeBoxModule } from 'devextreme-angular/ui/date-range-box';
export { DxDateRangeBoxComponent, DxDateRangeBoxModule } from 'devextreme-angular/ui/date-range-box';
import { DxDeferRenderingModule } from 'devextreme-angular/ui/defer-rendering';
export { DxDeferRenderingComponent, DxDeferRenderingModule } from 'devextreme-angular/ui/defer-rendering';
import { DxDiagramModule } from 'devextreme-angular/ui/diagram';
export { DxDiagramComponent, DxDiagramModule } from 'devextreme-angular/ui/diagram';
import { DxDraggableModule } from 'devextreme-angular/ui/draggable';
export { DxDraggableComponent, DxDraggableModule } from 'devextreme-angular/ui/draggable';
import { DxDrawerModule } from 'devextreme-angular/ui/drawer';
export { DxDrawerComponent, DxDrawerModule } from 'devextreme-angular/ui/drawer';
import { DxDropDownBoxModule } from 'devextreme-angular/ui/drop-down-box';
export { DxDropDownBoxComponent, DxDropDownBoxModule } from 'devextreme-angular/ui/drop-down-box';
import { DxDropDownButtonModule } from 'devextreme-angular/ui/drop-down-button';
export { DxDropDownButtonComponent, DxDropDownButtonModule } from 'devextreme-angular/ui/drop-down-button';
import { DxFileManagerModule } from 'devextreme-angular/ui/file-manager';
export { DxFileManagerComponent, DxFileManagerModule } from 'devextreme-angular/ui/file-manager';
import { DxFileUploaderModule } from 'devextreme-angular/ui/file-uploader';
export { DxFileUploaderComponent, DxFileUploaderModule } from 'devextreme-angular/ui/file-uploader';
import { DxFilterBuilderModule } from 'devextreme-angular/ui/filter-builder';
export { DxFilterBuilderComponent, DxFilterBuilderModule } from 'devextreme-angular/ui/filter-builder';
import { DxFormModule } from 'devextreme-angular/ui/form';
export { DxFormComponent, DxFormModule } from 'devextreme-angular/ui/form';
import { DxFunnelModule } from 'devextreme-angular/ui/funnel';
export { DxFunnelComponent, DxFunnelModule } from 'devextreme-angular/ui/funnel';
import { DxGalleryModule } from 'devextreme-angular/ui/gallery';
export { DxGalleryComponent, DxGalleryModule } from 'devextreme-angular/ui/gallery';
import { DxGanttModule } from 'devextreme-angular/ui/gantt';
export { DxGanttComponent, DxGanttModule } from 'devextreme-angular/ui/gantt';
import { DxHtmlEditorModule } from 'devextreme-angular/ui/html-editor';
export { DxHtmlEditorComponent, DxHtmlEditorModule } from 'devextreme-angular/ui/html-editor';
import { DxLinearGaugeModule } from 'devextreme-angular/ui/linear-gauge';
export { DxLinearGaugeComponent, DxLinearGaugeModule } from 'devextreme-angular/ui/linear-gauge';
import { DxListModule } from 'devextreme-angular/ui/list';
export { DxListComponent, DxListModule } from 'devextreme-angular/ui/list';
import { DxLoadIndicatorModule } from 'devextreme-angular/ui/load-indicator';
export { DxLoadIndicatorComponent, DxLoadIndicatorModule } from 'devextreme-angular/ui/load-indicator';
import { DxLoadPanelModule } from 'devextreme-angular/ui/load-panel';
export { DxLoadPanelComponent, DxLoadPanelModule } from 'devextreme-angular/ui/load-panel';
import { DxLookupModule } from 'devextreme-angular/ui/lookup';
export { DxLookupComponent, DxLookupModule } from 'devextreme-angular/ui/lookup';
import { DxMapModule } from 'devextreme-angular/ui/map';
export { DxMapComponent, DxMapModule } from 'devextreme-angular/ui/map';
import { DxMenuModule } from 'devextreme-angular/ui/menu';
export { DxMenuComponent, DxMenuModule } from 'devextreme-angular/ui/menu';
import { DxMultiViewModule } from 'devextreme-angular/ui/multi-view';
export { DxMultiViewComponent, DxMultiViewModule } from 'devextreme-angular/ui/multi-view';
import { DxNumberBoxModule } from 'devextreme-angular/ui/number-box';
export { DxNumberBoxComponent, DxNumberBoxModule } from 'devextreme-angular/ui/number-box';
import { DxPieChartModule } from 'devextreme-angular/ui/pie-chart';
export { DxPieChartComponent, DxPieChartModule } from 'devextreme-angular/ui/pie-chart';
import { DxPivotGridModule } from 'devextreme-angular/ui/pivot-grid';
export { DxPivotGridComponent, DxPivotGridModule } from 'devextreme-angular/ui/pivot-grid';
import { DxPivotGridFieldChooserModule } from 'devextreme-angular/ui/pivot-grid-field-chooser';
export { DxPivotGridFieldChooserComponent, DxPivotGridFieldChooserModule } from 'devextreme-angular/ui/pivot-grid-field-chooser';
import { DxPolarChartModule } from 'devextreme-angular/ui/polar-chart';
export { DxPolarChartComponent, DxPolarChartModule } from 'devextreme-angular/ui/polar-chart';
import { DxPopoverModule } from 'devextreme-angular/ui/popover';
export { DxPopoverComponent, DxPopoverModule } from 'devextreme-angular/ui/popover';
import { DxPopupModule } from 'devextreme-angular/ui/popup';
export { DxPopupComponent, DxPopupModule } from 'devextreme-angular/ui/popup';
import { DxProgressBarModule } from 'devextreme-angular/ui/progress-bar';
export { DxProgressBarComponent, DxProgressBarModule } from 'devextreme-angular/ui/progress-bar';
import { DxRadioGroupModule } from 'devextreme-angular/ui/radio-group';
export { DxRadioGroupComponent, DxRadioGroupModule } from 'devextreme-angular/ui/radio-group';
import { DxRangeSelectorModule } from 'devextreme-angular/ui/range-selector';
export { DxRangeSelectorComponent, DxRangeSelectorModule } from 'devextreme-angular/ui/range-selector';
import { DxRangeSliderModule } from 'devextreme-angular/ui/range-slider';
export { DxRangeSliderComponent, DxRangeSliderModule } from 'devextreme-angular/ui/range-slider';
import { DxRecurrenceEditorModule } from 'devextreme-angular/ui/recurrence-editor';
export { DxRecurrenceEditorComponent, DxRecurrenceEditorModule } from 'devextreme-angular/ui/recurrence-editor';
import { DxResizableModule } from 'devextreme-angular/ui/resizable';
export { DxResizableComponent, DxResizableModule } from 'devextreme-angular/ui/resizable';
import { DxResponsiveBoxModule } from 'devextreme-angular/ui/responsive-box';
export { DxResponsiveBoxComponent, DxResponsiveBoxModule } from 'devextreme-angular/ui/responsive-box';
import { DxSankeyModule } from 'devextreme-angular/ui/sankey';
export { DxSankeyComponent, DxSankeyModule } from 'devextreme-angular/ui/sankey';
import { DxSchedulerModule } from 'devextreme-angular/ui/scheduler';
export { DxSchedulerComponent, DxSchedulerModule } from 'devextreme-angular/ui/scheduler';
import { DxScrollViewModule } from 'devextreme-angular/ui/scroll-view';
export { DxScrollViewComponent, DxScrollViewModule } from 'devextreme-angular/ui/scroll-view';
import { DxSelectBoxModule } from 'devextreme-angular/ui/select-box';
export { DxSelectBoxComponent, DxSelectBoxModule } from 'devextreme-angular/ui/select-box';
import { DxSliderModule } from 'devextreme-angular/ui/slider';
export { DxSliderComponent, DxSliderModule } from 'devextreme-angular/ui/slider';
import { DxSortableModule } from 'devextreme-angular/ui/sortable';
export { DxSortableComponent, DxSortableModule } from 'devextreme-angular/ui/sortable';
import { DxSparklineModule } from 'devextreme-angular/ui/sparkline';
export { DxSparklineComponent, DxSparklineModule } from 'devextreme-angular/ui/sparkline';
import { DxSpeedDialActionModule } from 'devextreme-angular/ui/speed-dial-action';
export { DxSpeedDialActionComponent, DxSpeedDialActionModule } from 'devextreme-angular/ui/speed-dial-action';
import { DxSwitchModule } from 'devextreme-angular/ui/switch';
export { DxSwitchComponent, DxSwitchModule } from 'devextreme-angular/ui/switch';
import { DxTabPanelModule } from 'devextreme-angular/ui/tab-panel';
export { DxTabPanelComponent, DxTabPanelModule } from 'devextreme-angular/ui/tab-panel';
import { DxTabsModule } from 'devextreme-angular/ui/tabs';
export { DxTabsComponent, DxTabsModule } from 'devextreme-angular/ui/tabs';
import { DxTagBoxModule } from 'devextreme-angular/ui/tag-box';
export { DxTagBoxComponent, DxTagBoxModule } from 'devextreme-angular/ui/tag-box';
import { DxTextAreaModule } from 'devextreme-angular/ui/text-area';
export { DxTextAreaComponent, DxTextAreaModule } from 'devextreme-angular/ui/text-area';
import { DxTextBoxModule } from 'devextreme-angular/ui/text-box';
export { DxTextBoxComponent, DxTextBoxModule } from 'devextreme-angular/ui/text-box';
import { DxTileViewModule } from 'devextreme-angular/ui/tile-view';
export { DxTileViewComponent, DxTileViewModule } from 'devextreme-angular/ui/tile-view';
import { DxToastModule } from 'devextreme-angular/ui/toast';
export { DxToastComponent, DxToastModule } from 'devextreme-angular/ui/toast';
import { DxToolbarModule } from 'devextreme-angular/ui/toolbar';
export { DxToolbarComponent, DxToolbarModule } from 'devextreme-angular/ui/toolbar';
import { DxTooltipModule } from 'devextreme-angular/ui/tooltip';
export { DxTooltipComponent, DxTooltipModule } from 'devextreme-angular/ui/tooltip';
import { DxTreeListModule } from 'devextreme-angular/ui/tree-list';
export { DxTreeListComponent, DxTreeListModule } from 'devextreme-angular/ui/tree-list';
import { DxTreeMapModule } from 'devextreme-angular/ui/tree-map';
export { DxTreeMapComponent, DxTreeMapModule } from 'devextreme-angular/ui/tree-map';
import { DxTreeViewModule } from 'devextreme-angular/ui/tree-view';
export { DxTreeViewComponent, DxTreeViewModule } from 'devextreme-angular/ui/tree-view';
import { DxValidationGroupModule } from 'devextreme-angular/ui/validation-group';
export { DxValidationGroupComponent, DxValidationGroupModule } from 'devextreme-angular/ui/validation-group';
import { DxValidationSummaryModule } from 'devextreme-angular/ui/validation-summary';
export { DxValidationSummaryComponent, DxValidationSummaryModule } from 'devextreme-angular/ui/validation-summary';
import { DxValidatorModule } from 'devextreme-angular/ui/validator';
export { DxValidatorComponent, DxValidatorModule } from 'devextreme-angular/ui/validator';
import { DxVectorMapModule } from 'devextreme-angular/ui/vector-map';
export { DxVectorMapComponent, DxVectorMapModule } from 'devextreme-angular/ui/vector-map';
import 'devextreme/common/charts';

/*!
 * devextreme-angular
 * Version: 23.2.4
 * Build date: Mon Jan 29 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
class DevExtremeModule {
}
/** @nocollapse */ DevExtremeModule.ɵfac = function DevExtremeModule_Factory(t) { return new (t || DevExtremeModule)(); };
/** @nocollapse */ DevExtremeModule.ɵmod = /** @pureOrBreakMyCode */ i0.ɵɵdefineNgModule({ type: DevExtremeModule });
/** @nocollapse */ DevExtremeModule.ɵinj = /** @pureOrBreakMyCode */ i0.ɵɵdefineInjector({ imports: [[
            DxAccordionModule,
            DxActionSheetModule,
            DxAutocompleteModule,
            DxBarGaugeModule,
            DxBoxModule,
            DxBulletModule,
            DxButtonModule,
            DxButtonGroupModule,
            DxCalendarModule,
            DxChartModule,
            DxCheckBoxModule,
            DxCircularGaugeModule,
            DxColorBoxModule,
            DxContextMenuModule,
            DxDataGridModule,
            DxDateBoxModule,
            DxDateRangeBoxModule,
            DxDeferRenderingModule,
            DxDiagramModule,
            DxDraggableModule,
            DxDrawerModule,
            DxDropDownBoxModule,
            DxDropDownButtonModule,
            DxFileManagerModule,
            DxFileUploaderModule,
            DxFilterBuilderModule,
            DxFormModule,
            DxFunnelModule,
            DxGalleryModule,
            DxGanttModule,
            DxHtmlEditorModule,
            DxLinearGaugeModule,
            DxListModule,
            DxLoadIndicatorModule,
            DxLoadPanelModule,
            DxLookupModule,
            DxMapModule,
            DxMenuModule,
            DxMultiViewModule,
            DxNumberBoxModule,
            DxPieChartModule,
            DxPivotGridModule,
            DxPivotGridFieldChooserModule,
            DxPolarChartModule,
            DxPopoverModule,
            DxPopupModule,
            DxProgressBarModule,
            DxRadioGroupModule,
            DxRangeSelectorModule,
            DxRangeSliderModule,
            DxRecurrenceEditorModule,
            DxResizableModule,
            DxResponsiveBoxModule,
            DxSankeyModule,
            DxSchedulerModule,
            DxScrollViewModule,
            DxSelectBoxModule,
            DxSliderModule,
            DxSortableModule,
            DxSparklineModule,
            DxSpeedDialActionModule,
            DxSwitchModule,
            DxTabPanelModule,
            DxTabsModule,
            DxTagBoxModule,
            DxTextAreaModule,
            DxTextBoxModule,
            DxTileViewModule,
            DxToastModule,
            DxToolbarModule,
            DxTooltipModule,
            DxTreeListModule,
            DxTreeMapModule,
            DxTreeViewModule,
            DxValidationGroupModule,
            DxValidationSummaryModule,
            DxValidatorModule,
            DxVectorMapModule,
            DxTemplateModule
        ], DxAccordionModule,
        DxActionSheetModule,
        DxAutocompleteModule,
        DxBarGaugeModule,
        DxBoxModule,
        DxBulletModule,
        DxButtonModule,
        DxButtonGroupModule,
        DxCalendarModule,
        DxChartModule,
        DxCheckBoxModule,
        DxCircularGaugeModule,
        DxColorBoxModule,
        DxContextMenuModule,
        DxDataGridModule,
        DxDateBoxModule,
        DxDateRangeBoxModule,
        DxDeferRenderingModule,
        DxDiagramModule,
        DxDraggableModule,
        DxDrawerModule,
        DxDropDownBoxModule,
        DxDropDownButtonModule,
        DxFileManagerModule,
        DxFileUploaderModule,
        DxFilterBuilderModule,
        DxFormModule,
        DxFunnelModule,
        DxGalleryModule,
        DxGanttModule,
        DxHtmlEditorModule,
        DxLinearGaugeModule,
        DxListModule,
        DxLoadIndicatorModule,
        DxLoadPanelModule,
        DxLookupModule,
        DxMapModule,
        DxMenuModule,
        DxMultiViewModule,
        DxNumberBoxModule,
        DxPieChartModule,
        DxPivotGridModule,
        DxPivotGridFieldChooserModule,
        DxPolarChartModule,
        DxPopoverModule,
        DxPopupModule,
        DxProgressBarModule,
        DxRadioGroupModule,
        DxRangeSelectorModule,
        DxRangeSliderModule,
        DxRecurrenceEditorModule,
        DxResizableModule,
        DxResponsiveBoxModule,
        DxSankeyModule,
        DxSchedulerModule,
        DxScrollViewModule,
        DxSelectBoxModule,
        DxSliderModule,
        DxSortableModule,
        DxSparklineModule,
        DxSpeedDialActionModule,
        DxSwitchModule,
        DxTabPanelModule,
        DxTabsModule,
        DxTagBoxModule,
        DxTextAreaModule,
        DxTextBoxModule,
        DxTileViewModule,
        DxToastModule,
        DxToolbarModule,
        DxTooltipModule,
        DxTreeListModule,
        DxTreeMapModule,
        DxTreeViewModule,
        DxValidationGroupModule,
        DxValidationSummaryModule,
        DxValidatorModule,
        DxVectorMapModule,
        DxTemplateModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DevExtremeModule, [{
        type: NgModule,
        args: [{
                imports: [
                    DxAccordionModule,
                    DxActionSheetModule,
                    DxAutocompleteModule,
                    DxBarGaugeModule,
                    DxBoxModule,
                    DxBulletModule,
                    DxButtonModule,
                    DxButtonGroupModule,
                    DxCalendarModule,
                    DxChartModule,
                    DxCheckBoxModule,
                    DxCircularGaugeModule,
                    DxColorBoxModule,
                    DxContextMenuModule,
                    DxDataGridModule,
                    DxDateBoxModule,
                    DxDateRangeBoxModule,
                    DxDeferRenderingModule,
                    DxDiagramModule,
                    DxDraggableModule,
                    DxDrawerModule,
                    DxDropDownBoxModule,
                    DxDropDownButtonModule,
                    DxFileManagerModule,
                    DxFileUploaderModule,
                    DxFilterBuilderModule,
                    DxFormModule,
                    DxFunnelModule,
                    DxGalleryModule,
                    DxGanttModule,
                    DxHtmlEditorModule,
                    DxLinearGaugeModule,
                    DxListModule,
                    DxLoadIndicatorModule,
                    DxLoadPanelModule,
                    DxLookupModule,
                    DxMapModule,
                    DxMenuModule,
                    DxMultiViewModule,
                    DxNumberBoxModule,
                    DxPieChartModule,
                    DxPivotGridModule,
                    DxPivotGridFieldChooserModule,
                    DxPolarChartModule,
                    DxPopoverModule,
                    DxPopupModule,
                    DxProgressBarModule,
                    DxRadioGroupModule,
                    DxRangeSelectorModule,
                    DxRangeSliderModule,
                    DxRecurrenceEditorModule,
                    DxResizableModule,
                    DxResponsiveBoxModule,
                    DxSankeyModule,
                    DxSchedulerModule,
                    DxScrollViewModule,
                    DxSelectBoxModule,
                    DxSliderModule,
                    DxSortableModule,
                    DxSparklineModule,
                    DxSpeedDialActionModule,
                    DxSwitchModule,
                    DxTabPanelModule,
                    DxTabsModule,
                    DxTagBoxModule,
                    DxTextAreaModule,
                    DxTextBoxModule,
                    DxTileViewModule,
                    DxToastModule,
                    DxToolbarModule,
                    DxTooltipModule,
                    DxTreeListModule,
                    DxTreeMapModule,
                    DxTreeViewModule,
                    DxValidationGroupModule,
                    DxValidationSummaryModule,
                    DxValidatorModule,
                    DxVectorMapModule,
                    DxTemplateModule
                ],
                exports: [
                    DxAccordionModule,
                    DxActionSheetModule,
                    DxAutocompleteModule,
                    DxBarGaugeModule,
                    DxBoxModule,
                    DxBulletModule,
                    DxButtonModule,
                    DxButtonGroupModule,
                    DxCalendarModule,
                    DxChartModule,
                    DxCheckBoxModule,
                    DxCircularGaugeModule,
                    DxColorBoxModule,
                    DxContextMenuModule,
                    DxDataGridModule,
                    DxDateBoxModule,
                    DxDateRangeBoxModule,
                    DxDeferRenderingModule,
                    DxDiagramModule,
                    DxDraggableModule,
                    DxDrawerModule,
                    DxDropDownBoxModule,
                    DxDropDownButtonModule,
                    DxFileManagerModule,
                    DxFileUploaderModule,
                    DxFilterBuilderModule,
                    DxFormModule,
                    DxFunnelModule,
                    DxGalleryModule,
                    DxGanttModule,
                    DxHtmlEditorModule,
                    DxLinearGaugeModule,
                    DxListModule,
                    DxLoadIndicatorModule,
                    DxLoadPanelModule,
                    DxLookupModule,
                    DxMapModule,
                    DxMenuModule,
                    DxMultiViewModule,
                    DxNumberBoxModule,
                    DxPieChartModule,
                    DxPivotGridModule,
                    DxPivotGridFieldChooserModule,
                    DxPolarChartModule,
                    DxPopoverModule,
                    DxPopupModule,
                    DxProgressBarModule,
                    DxRadioGroupModule,
                    DxRangeSelectorModule,
                    DxRangeSliderModule,
                    DxRecurrenceEditorModule,
                    DxResizableModule,
                    DxResponsiveBoxModule,
                    DxSankeyModule,
                    DxSchedulerModule,
                    DxScrollViewModule,
                    DxSelectBoxModule,
                    DxSliderModule,
                    DxSortableModule,
                    DxSparklineModule,
                    DxSpeedDialActionModule,
                    DxSwitchModule,
                    DxTabPanelModule,
                    DxTabsModule,
                    DxTagBoxModule,
                    DxTextAreaModule,
                    DxTextBoxModule,
                    DxTileViewModule,
                    DxToastModule,
                    DxToolbarModule,
                    DxTooltipModule,
                    DxTreeListModule,
                    DxTreeMapModule,
                    DxTreeViewModule,
                    DxValidationGroupModule,
                    DxValidationSummaryModule,
                    DxValidatorModule,
                    DxVectorMapModule,
                    DxTemplateModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(DevExtremeModule, { imports: [DxAccordionModule,
        DxActionSheetModule,
        DxAutocompleteModule,
        DxBarGaugeModule,
        DxBoxModule,
        DxBulletModule,
        DxButtonModule,
        DxButtonGroupModule,
        DxCalendarModule,
        DxChartModule,
        DxCheckBoxModule,
        DxCircularGaugeModule,
        DxColorBoxModule,
        DxContextMenuModule,
        DxDataGridModule,
        DxDateBoxModule,
        DxDateRangeBoxModule,
        DxDeferRenderingModule,
        DxDiagramModule,
        DxDraggableModule,
        DxDrawerModule,
        DxDropDownBoxModule,
        DxDropDownButtonModule,
        DxFileManagerModule,
        DxFileUploaderModule,
        DxFilterBuilderModule,
        DxFormModule,
        DxFunnelModule,
        DxGalleryModule,
        DxGanttModule,
        DxHtmlEditorModule,
        DxLinearGaugeModule,
        DxListModule,
        DxLoadIndicatorModule,
        DxLoadPanelModule,
        DxLookupModule,
        DxMapModule,
        DxMenuModule,
        DxMultiViewModule,
        DxNumberBoxModule,
        DxPieChartModule,
        DxPivotGridModule,
        DxPivotGridFieldChooserModule,
        DxPolarChartModule,
        DxPopoverModule,
        DxPopupModule,
        DxProgressBarModule,
        DxRadioGroupModule,
        DxRangeSelectorModule,
        DxRangeSliderModule,
        DxRecurrenceEditorModule,
        DxResizableModule,
        DxResponsiveBoxModule,
        DxSankeyModule,
        DxSchedulerModule,
        DxScrollViewModule,
        DxSelectBoxModule,
        DxSliderModule,
        DxSortableModule,
        DxSparklineModule,
        DxSpeedDialActionModule,
        DxSwitchModule,
        DxTabPanelModule,
        DxTabsModule,
        DxTagBoxModule,
        DxTextAreaModule,
        DxTextBoxModule,
        DxTileViewModule,
        DxToastModule,
        DxToolbarModule,
        DxTooltipModule,
        DxTreeListModule,
        DxTreeMapModule,
        DxTreeViewModule,
        DxValidationGroupModule,
        DxValidationSummaryModule,
        DxValidatorModule,
        DxVectorMapModule,
        DxTemplateModule], exports: [DxAccordionModule,
        DxActionSheetModule,
        DxAutocompleteModule,
        DxBarGaugeModule,
        DxBoxModule,
        DxBulletModule,
        DxButtonModule,
        DxButtonGroupModule,
        DxCalendarModule,
        DxChartModule,
        DxCheckBoxModule,
        DxCircularGaugeModule,
        DxColorBoxModule,
        DxContextMenuModule,
        DxDataGridModule,
        DxDateBoxModule,
        DxDateRangeBoxModule,
        DxDeferRenderingModule,
        DxDiagramModule,
        DxDraggableModule,
        DxDrawerModule,
        DxDropDownBoxModule,
        DxDropDownButtonModule,
        DxFileManagerModule,
        DxFileUploaderModule,
        DxFilterBuilderModule,
        DxFormModule,
        DxFunnelModule,
        DxGalleryModule,
        DxGanttModule,
        DxHtmlEditorModule,
        DxLinearGaugeModule,
        DxListModule,
        DxLoadIndicatorModule,
        DxLoadPanelModule,
        DxLookupModule,
        DxMapModule,
        DxMenuModule,
        DxMultiViewModule,
        DxNumberBoxModule,
        DxPieChartModule,
        DxPivotGridModule,
        DxPivotGridFieldChooserModule,
        DxPolarChartModule,
        DxPopoverModule,
        DxPopupModule,
        DxProgressBarModule,
        DxRadioGroupModule,
        DxRangeSelectorModule,
        DxRangeSliderModule,
        DxRecurrenceEditorModule,
        DxResizableModule,
        DxResponsiveBoxModule,
        DxSankeyModule,
        DxSchedulerModule,
        DxScrollViewModule,
        DxSelectBoxModule,
        DxSliderModule,
        DxSortableModule,
        DxSparklineModule,
        DxSpeedDialActionModule,
        DxSwitchModule,
        DxTabPanelModule,
        DxTabsModule,
        DxTagBoxModule,
        DxTextAreaModule,
        DxTextBoxModule,
        DxTileViewModule,
        DxToastModule,
        DxToolbarModule,
        DxTooltipModule,
        DxTreeListModule,
        DxTreeMapModule,
        DxTreeViewModule,
        DxValidationGroupModule,
        DxValidationSummaryModule,
        DxValidatorModule,
        DxVectorMapModule,
        DxTemplateModule] }); })();

/*!
 * devextreme-angular
 * Version: 23.2.4
 * Build date: Mon Jan 29 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */

/*!
 * devextreme-angular
 * Version: 23.2.4
 * Build date: Mon Jan 29 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */

/*!
 * devextreme-angular
 * Version: 23.2.4
 * Build date: Mon Jan 29 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */

/*!
 * devextreme-angular
 * Version: 23.2.4
 * Build date: Mon Jan 29 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */

/**
 * Generated bundle index. Do not edit.
 */

export { DevExtremeModule };
