// devextreme-angular/bundles/devextreme-angular-server.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('devextreme/core/inferno_renderer'), require('inferno-server')) :
    typeof define === 'function' && define.amd ? define('devextreme-angular/server', ['exports', '@angular/core', '@angular/common', 'devextreme/core/inferno_renderer', 'inferno-server'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["devextreme-angular"] = global["devextreme-angular"] || {}, global["devextreme-angular"].server = {}), global.ng.core, global.ng.common, global.infernoRenderer, global.infernoServer));
})(this, (function (exports, i0, common, infernoRenderer, infernoServer) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var infernoRenderer__default = /*#__PURE__*/_interopDefaultLegacy(infernoRenderer);

    /*!
     * devextreme-angular
     * Version: 23.2.4
     * Build date: Mon Jan 29 2024
     *
     * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
     *
     * This software may be modified and distributed under the terms
     * of the MIT license. See the LICENSE file in the root of the project for details.
     *
     * https://github.com/DevExpress/devextreme-angular
     */
    var DxServerModule = /** @class */ (function () {
        function DxServerModule(platformId) {
            if (common.isPlatformServer(platformId)) {
                infernoRenderer__default["default"].inject({
                    render: function (component, props, container) {
                        var el = infernoRenderer__default["default"].createElement(component, props);
                        var document = container.ownerDocument;
                        var temp = document.createElement(container.tagName);
                        temp.innerHTML = infernoServer.renderToString(el);
                        var mainElement = temp.childNodes[0];
                        var childString = mainElement.innerHTML;
                        for (var i = 0; i < mainElement.attributes.length; i++) {
                            temp.setAttribute(mainElement.attributes[i].name, mainElement.attributes[i].value);
                        }
                        temp.innerHTML = childString;
                        container.outerHTML = temp.outerHTML;
                    },
                });
            }
        }
        return DxServerModule;
    }());
    /** @nocollapse */ DxServerModule.ɵfac = function DxServerModule_Factory(t) { return new (t || DxServerModule)(i0__namespace.ɵɵinject(i0.PLATFORM_ID)); };
    /** @nocollapse */ DxServerModule.ɵmod = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineNgModule({ type: DxServerModule });
    /** @nocollapse */ DxServerModule.ɵinj = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineInjector({ providers: [], imports: [[]] });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.ɵsetClassMetadata(DxServerModule, [{
                type: i0.NgModule,
                args: [{
                        exports: [],
                        imports: [],
                        providers: [],
                    }]
            }], function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i0.PLATFORM_ID]
                        }] }];
        }, null);
    })();

    /*!
     * devextreme-angular
     * Version: 23.2.4
     * Build date: Mon Jan 29 2024
     *
     * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
     *
     * This software may be modified and distributed under the terms
     * of the MIT license. See the LICENSE file in the root of the project for details.
     *
     * https://github.com/DevExpress/devextreme-angular
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.DxServerModule = DxServerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
