// devextreme-angular/bundles/devextreme-angular-ui-chart.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('devextreme/viz/chart'), require('devextreme-angular/core'), require('devextreme-angular/ui/nested'), require('@angular/platform-browser')) :
  typeof define === 'function' && define.amd ? define('devextreme-angular/ui/chart', ['exports', '@angular/core', 'devextreme/viz/chart', 'devextreme-angular/core', 'devextreme-angular/ui/nested', '@angular/platform-browser'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["devextreme-angular"] = global["devextreme-angular"] || {}, global["devextreme-angular"].ui = global["devextreme-angular"].ui || {}, global["devextreme-angular"].ui.chart = {}), global.ng.core, global.DxChart, global["devextreme-angular"].core, global["devextreme-angular"].ui.nested, global.ng.platformBrowser));
})(this, (function (exports, i0, DxChart, i1, nested, i2) { 'use strict';

  function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

  function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
      Object.keys(e).forEach(function (k) {
        if (k !== 'default') {
          var d = Object.getOwnPropertyDescriptor(e, k);
          Object.defineProperty(n, k, d.get ? d : {
            enumerable: true,
            get: function () { return e[k]; }
          });
        }
      });
    }
    n["default"] = e;
    return Object.freeze(n);
  }

  var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
  var DxChart__default = /*#__PURE__*/_interopDefaultLegacy(DxChart);
  var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
  var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

  /******************************************************************************
  Copyright (c) Microsoft Corporation.

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted.

  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
  REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
  AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
  LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  PERFORMANCE OF THIS SOFTWARE.
  ***************************************************************************** */
  /* global Reflect, Promise, SuppressedError, Symbol */
  var extendStatics = function (d, b) {
      extendStatics = Object.setPrototypeOf ||
          ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
          function (d, b) { for (var p in b)
              if (Object.prototype.hasOwnProperty.call(b, p))
                  d[p] = b[p]; };
      return extendStatics(d, b);
  };
  function __extends(d, b) {
      if (typeof b !== "function" && b !== null)
          throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
      extendStatics(d, b);
      function __() { this.constructor = d; }
      d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  }
  var __assign = function () {
      __assign = Object.assign || function __assign(t) {
          for (var s, i = 1, n = arguments.length; i < n; i++) {
              s = arguments[i];
              for (var p in s)
                  if (Object.prototype.hasOwnProperty.call(s, p))
                      t[p] = s[p];
          }
          return t;
      };
      return __assign.apply(this, arguments);
  };
  function __rest(s, e) {
      var t = {};
      for (var p in s)
          if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
              t[p] = s[p];
      if (s != null && typeof Object.getOwnPropertySymbols === "function")
          for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
              if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                  t[p[i]] = s[p[i]];
          }
      return t;
  }
  function __decorate(decorators, target, key, desc) {
      var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
      if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
          r = Reflect.decorate(decorators, target, key, desc);
      else
          for (var i = decorators.length - 1; i >= 0; i--)
              if (d = decorators[i])
                  r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
      return c > 3 && r && Object.defineProperty(target, key, r), r;
  }
  function __param(paramIndex, decorator) {
      return function (target, key) { decorator(target, key, paramIndex); };
  }
  function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
      function accept(f) { if (f !== void 0 && typeof f !== "function")
          throw new TypeError("Function expected"); return f; }
      var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
      var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
      var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
      var _, done = false;
      for (var i = decorators.length - 1; i >= 0; i--) {
          var context = {};
          for (var p in contextIn)
              context[p] = p === "access" ? {} : contextIn[p];
          for (var p in contextIn.access)
              context.access[p] = contextIn.access[p];
          context.addInitializer = function (f) { if (done)
              throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
          var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
          if (kind === "accessor") {
              if (result === void 0)
                  continue;
              if (result === null || typeof result !== "object")
                  throw new TypeError("Object expected");
              if (_ = accept(result.get))
                  descriptor.get = _;
              if (_ = accept(result.set))
                  descriptor.set = _;
              if (_ = accept(result.init))
                  initializers.unshift(_);
          }
          else if (_ = accept(result)) {
              if (kind === "field")
                  initializers.unshift(_);
              else
                  descriptor[key] = _;
          }
      }
      if (target)
          Object.defineProperty(target, contextIn.name, descriptor);
      done = true;
  }
  ;
  function __runInitializers(thisArg, initializers, value) {
      var useValue = arguments.length > 2;
      for (var i = 0; i < initializers.length; i++) {
          value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
      }
      return useValue ? value : void 0;
  }
  ;
  function __propKey(x) {
      return typeof x === "symbol" ? x : "".concat(x);
  }
  ;
  function __setFunctionName(f, name, prefix) {
      if (typeof name === "symbol")
          name = name.description ? "[".concat(name.description, "]") : "";
      return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
  }
  ;
  function __metadata(metadataKey, metadataValue) {
      if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
          return Reflect.metadata(metadataKey, metadataValue);
  }
  function __awaiter(thisArg, _arguments, P, generator) {
      function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
      return new (P || (P = Promise))(function (resolve, reject) {
          function fulfilled(value) { try {
              step(generator.next(value));
          }
          catch (e) {
              reject(e);
          } }
          function rejected(value) { try {
              step(generator["throw"](value));
          }
          catch (e) {
              reject(e);
          } }
          function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
          step((generator = generator.apply(thisArg, _arguments || [])).next());
      });
  }
  function __generator(thisArg, body) {
      var _ = { label: 0, sent: function () { if (t[0] & 1)
              throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
      return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
      function verb(n) { return function (v) { return step([n, v]); }; }
      function step(op) {
          if (f)
              throw new TypeError("Generator is already executing.");
          while (g && (g = 0, op[0] && (_ = 0)), _)
              try {
                  if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                      return t;
                  if (y = 0, t)
                      op = [op[0] & 2, t.value];
                  switch (op[0]) {
                      case 0:
                      case 1:
                          t = op;
                          break;
                      case 4:
                          _.label++;
                          return { value: op[1], done: false };
                      case 5:
                          _.label++;
                          y = op[1];
                          op = [0];
                          continue;
                      case 7:
                          op = _.ops.pop();
                          _.trys.pop();
                          continue;
                      default:
                          if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                              _ = 0;
                              continue;
                          }
                          if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                              _.label = op[1];
                              break;
                          }
                          if (op[0] === 6 && _.label < t[1]) {
                              _.label = t[1];
                              t = op;
                              break;
                          }
                          if (t && _.label < t[2]) {
                              _.label = t[2];
                              _.ops.push(op);
                              break;
                          }
                          if (t[2])
                              _.ops.pop();
                          _.trys.pop();
                          continue;
                  }
                  op = body.call(thisArg, _);
              }
              catch (e) {
                  op = [6, e];
                  y = 0;
              }
              finally {
                  f = t = 0;
              }
          if (op[0] & 5)
              throw op[1];
          return { value: op[0] ? op[1] : void 0, done: true };
      }
  }
  var __createBinding = Object.create ? (function (o, m, k, k2) {
      if (k2 === undefined)
          k2 = k;
      var desc = Object.getOwnPropertyDescriptor(m, k);
      if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
          desc = { enumerable: true, get: function () { return m[k]; } };
      }
      Object.defineProperty(o, k2, desc);
  }) : (function (o, m, k, k2) {
      if (k2 === undefined)
          k2 = k;
      o[k2] = m[k];
  });
  function __exportStar(m, o) {
      for (var p in m)
          if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
              __createBinding(o, m, p);
  }
  function __values(o) {
      var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
      if (m)
          return m.call(o);
      if (o && typeof o.length === "number")
          return {
              next: function () {
                  if (o && i >= o.length)
                      o = void 0;
                  return { value: o && o[i++], done: !o };
              }
          };
      throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
  }
  function __read(o, n) {
      var m = typeof Symbol === "function" && o[Symbol.iterator];
      if (!m)
          return o;
      var i = m.call(o), r, ar = [], e;
      try {
          while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
              ar.push(r.value);
      }
      catch (error) {
          e = { error: error };
      }
      finally {
          try {
              if (r && !r.done && (m = i["return"]))
                  m.call(i);
          }
          finally {
              if (e)
                  throw e.error;
          }
      }
      return ar;
  }
  /** @deprecated */
  function __spread() {
      for (var ar = [], i = 0; i < arguments.length; i++)
          ar = ar.concat(__read(arguments[i]));
      return ar;
  }
  /** @deprecated */
  function __spreadArrays() {
      for (var s = 0, i = 0, il = arguments.length; i < il; i++)
          s += arguments[i].length;
      for (var r = Array(s), k = 0, i = 0; i < il; i++)
          for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
              r[k] = a[j];
      return r;
  }
  function __spreadArray(to, from, pack) {
      if (pack || arguments.length === 2)
          for (var i = 0, l = from.length, ar; i < l; i++) {
              if (ar || !(i in from)) {
                  if (!ar)
                      ar = Array.prototype.slice.call(from, 0, i);
                  ar[i] = from[i];
              }
          }
      return to.concat(ar || Array.prototype.slice.call(from));
  }
  function __await(v) {
      return this instanceof __await ? (this.v = v, this) : new __await(v);
  }
  function __asyncGenerator(thisArg, _arguments, generator) {
      if (!Symbol.asyncIterator)
          throw new TypeError("Symbol.asyncIterator is not defined.");
      var g = generator.apply(thisArg, _arguments || []), i, q = [];
      return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
      function verb(n) { if (g[n])
          i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
      function resume(n, v) { try {
          step(g[n](v));
      }
      catch (e) {
          settle(q[0][3], e);
      } }
      function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
      function fulfill(value) { resume("next", value); }
      function reject(value) { resume("throw", value); }
      function settle(f, v) { if (f(v), q.shift(), q.length)
          resume(q[0][0], q[0][1]); }
  }
  function __asyncDelegator(o) {
      var i, p;
      return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
      function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
  }
  function __asyncValues(o) {
      if (!Symbol.asyncIterator)
          throw new TypeError("Symbol.asyncIterator is not defined.");
      var m = o[Symbol.asyncIterator], i;
      return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
      function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
      function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
  }
  function __makeTemplateObject(cooked, raw) {
      if (Object.defineProperty) {
          Object.defineProperty(cooked, "raw", { value: raw });
      }
      else {
          cooked.raw = raw;
      }
      return cooked;
  }
  ;
  var __setModuleDefault = Object.create ? (function (o, v) {
      Object.defineProperty(o, "default", { enumerable: true, value: v });
  }) : function (o, v) {
      o["default"] = v;
  };
  function __importStar(mod) {
      if (mod && mod.__esModule)
          return mod;
      var result = {};
      if (mod != null)
          for (var k in mod)
              if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                  __createBinding(result, mod, k);
      __setModuleDefault(result, mod);
      return result;
  }
  function __importDefault(mod) {
      return (mod && mod.__esModule) ? mod : { default: mod };
  }
  function __classPrivateFieldGet(receiver, state, kind, f) {
      if (kind === "a" && !f)
          throw new TypeError("Private accessor was defined without a getter");
      if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
          throw new TypeError("Cannot read private member from an object whose class did not declare it");
      return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
  }
  function __classPrivateFieldSet(receiver, state, value, kind, f) {
      if (kind === "m")
          throw new TypeError("Private method is not writable");
      if (kind === "a" && !f)
          throw new TypeError("Private accessor was defined without a setter");
      if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
          throw new TypeError("Cannot write private member to an object whose class did not declare it");
      return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
  }
  function __classPrivateFieldIn(state, receiver) {
      if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function"))
          throw new TypeError("Cannot use 'in' operator on non-object");
      return typeof state === "function" ? receiver === state : state.has(receiver);
  }
  function __addDisposableResource(env, value, async) {
      if (value !== null && value !== void 0) {
          if (typeof value !== "object" && typeof value !== "function")
              throw new TypeError("Object expected.");
          var dispose;
          if (async) {
              if (!Symbol.asyncDispose)
                  throw new TypeError("Symbol.asyncDispose is not defined.");
              dispose = value[Symbol.asyncDispose];
          }
          if (dispose === void 0) {
              if (!Symbol.dispose)
                  throw new TypeError("Symbol.dispose is not defined.");
              dispose = value[Symbol.dispose];
          }
          if (typeof dispose !== "function")
              throw new TypeError("Object not disposable.");
          env.stack.push({ value: value, dispose: dispose, async: async });
      }
      else if (async) {
          env.stack.push({ async: true });
      }
      return value;
  }
  var _SuppressedError = typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
      var e = new Error(message);
      return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
  };
  function __disposeResources(env) {
      function fail(e) {
          env.error = env.hasError ? new _SuppressedError(e, env.error, "An error was suppressed during disposal.") : e;
          env.hasError = true;
      }
      function next() {
          while (env.stack.length) {
              var rec = env.stack.pop();
              try {
                  var result = rec.dispose && rec.dispose.call(rec.value);
                  if (rec.async)
                      return Promise.resolve(result).then(next, function (e) { fail(e); return next(); });
              }
              catch (e) {
                  fail(e);
              }
          }
          if (env.hasError)
              throw env.error;
      }
      return next();
  }
  var tslib_es6 = {
      __extends: __extends,
      __assign: __assign,
      __rest: __rest,
      __decorate: __decorate,
      __param: __param,
      __metadata: __metadata,
      __awaiter: __awaiter,
      __generator: __generator,
      __createBinding: __createBinding,
      __exportStar: __exportStar,
      __values: __values,
      __read: __read,
      __spread: __spread,
      __spreadArrays: __spreadArrays,
      __spreadArray: __spreadArray,
      __await: __await,
      __asyncGenerator: __asyncGenerator,
      __asyncDelegator: __asyncDelegator,
      __asyncValues: __asyncValues,
      __makeTemplateObject: __makeTemplateObject,
      __importStar: __importStar,
      __importDefault: __importDefault,
      __classPrivateFieldGet: __classPrivateFieldGet,
      __classPrivateFieldSet: __classPrivateFieldSet,
      __classPrivateFieldIn: __classPrivateFieldIn,
      __addDisposableResource: __addDisposableResource,
      __disposeResources: __disposeResources,
  };

  /**
   * [descr:dxChart]

   */
  var DxChartComponent = /** @class */ (function (_super) {
      __extends(DxChartComponent, _super);
      function DxChartComponent(elementRef, ngZone, templateHost, _watcherHelper, _idh, optionHost, transferState, platformId) {
          var _this = _super.call(this, elementRef, ngZone, templateHost, _watcherHelper, transferState, platformId) || this;
          _this._watcherHelper = _watcherHelper;
          _this._idh = _idh;
          _this._createEventEmitters([
              { subscribe: 'argumentAxisClick', emit: 'onArgumentAxisClick' },
              { subscribe: 'disposing', emit: 'onDisposing' },
              { subscribe: 'done', emit: 'onDone' },
              { subscribe: 'drawn', emit: 'onDrawn' },
              { subscribe: 'exported', emit: 'onExported' },
              { subscribe: 'exporting', emit: 'onExporting' },
              { subscribe: 'fileSaving', emit: 'onFileSaving' },
              { subscribe: 'incidentOccurred', emit: 'onIncidentOccurred' },
              { subscribe: 'initialized', emit: 'onInitialized' },
              { subscribe: 'legendClick', emit: 'onLegendClick' },
              { subscribe: 'optionChanged', emit: 'onOptionChanged' },
              { subscribe: 'pointClick', emit: 'onPointClick' },
              { subscribe: 'pointHoverChanged', emit: 'onPointHoverChanged' },
              { subscribe: 'pointSelectionChanged', emit: 'onPointSelectionChanged' },
              { subscribe: 'seriesClick', emit: 'onSeriesClick' },
              { subscribe: 'seriesHoverChanged', emit: 'onSeriesHoverChanged' },
              { subscribe: 'seriesSelectionChanged', emit: 'onSeriesSelectionChanged' },
              { subscribe: 'tooltipHidden', emit: 'onTooltipHidden' },
              { subscribe: 'tooltipShown', emit: 'onTooltipShown' },
              { subscribe: 'zoomEnd', emit: 'onZoomEnd' },
              { subscribe: 'zoomStart', emit: 'onZoomStart' },
              { emit: 'adaptiveLayoutChange' },
              { emit: 'adjustOnZoomChange' },
              { emit: 'animationChange' },
              { emit: 'annotationsChange' },
              { emit: 'argumentAxisChange' },
              { emit: 'autoHidePointMarkersChange' },
              { emit: 'barGroupPaddingChange' },
              { emit: 'barGroupWidthChange' },
              { emit: 'commonAnnotationSettingsChange' },
              { emit: 'commonAxisSettingsChange' },
              { emit: 'commonPaneSettingsChange' },
              { emit: 'commonSeriesSettingsChange' },
              { emit: 'containerBackgroundColorChange' },
              { emit: 'crosshairChange' },
              { emit: 'customizeAnnotationChange' },
              { emit: 'customizeLabelChange' },
              { emit: 'customizePointChange' },
              { emit: 'dataPrepareSettingsChange' },
              { emit: 'dataSourceChange' },
              { emit: 'defaultPaneChange' },
              { emit: 'disabledChange' },
              { emit: 'elementAttrChange' },
              { emit: 'exportChange' },
              { emit: 'legendChange' },
              { emit: 'loadingIndicatorChange' },
              { emit: 'marginChange' },
              { emit: 'maxBubbleSizeChange' },
              { emit: 'minBubbleSizeChange' },
              { emit: 'negativesAsZeroesChange' },
              { emit: 'paletteChange' },
              { emit: 'paletteExtensionModeChange' },
              { emit: 'panesChange' },
              { emit: 'pathModifiedChange' },
              { emit: 'pointSelectionModeChange' },
              { emit: 'redrawOnResizeChange' },
              { emit: 'resizePanesOnZoomChange' },
              { emit: 'resolveLabelOverlappingChange' },
              { emit: 'rotatedChange' },
              { emit: 'rtlEnabledChange' },
              { emit: 'scrollBarChange' },
              { emit: 'seriesChange' },
              { emit: 'seriesSelectionModeChange' },
              { emit: 'seriesTemplateChange' },
              { emit: 'sizeChange' },
              { emit: 'stickyHoveringChange' },
              { emit: 'synchronizeMultiAxesChange' },
              { emit: 'themeChange' },
              { emit: 'titleChange' },
              { emit: 'tooltipChange' },
              { emit: 'valueAxisChange' },
              { emit: 'zoomAndPanChange' }
          ]);
          _this._idh.setHost(_this);
          optionHost.setHost(_this);
          return _this;
      }
      Object.defineProperty(DxChartComponent.prototype, "adaptiveLayout", {
          /**
           * [descr:BaseChartOptions.adaptiveLayout]
          
           */
          get: function () {
              return this._getOption('adaptiveLayout');
          },
          set: function (value) {
              this._setOption('adaptiveLayout', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "adjustOnZoom", {
          /**
           * [descr:dxChartOptions.adjustOnZoom]
          
           */
          get: function () {
              return this._getOption('adjustOnZoom');
          },
          set: function (value) {
              this._setOption('adjustOnZoom', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "animation", {
          /**
           * [descr:BaseChartOptions.animation]
          
           */
          get: function () {
              return this._getOption('animation');
          },
          set: function (value) {
              this._setOption('animation', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "annotations", {
          /**
           * [descr:dxChartOptions.annotations]
          
           */
          get: function () {
              return this._getOption('annotations');
          },
          set: function (value) {
              this._setOption('annotations', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "argumentAxis", {
          /**
           * [descr:dxChartOptions.argumentAxis]
          
           */
          get: function () {
              return this._getOption('argumentAxis');
          },
          set: function (value) {
              this._setOption('argumentAxis', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "autoHidePointMarkers", {
          /**
           * [descr:dxChartOptions.autoHidePointMarkers]
          
           */
          get: function () {
              return this._getOption('autoHidePointMarkers');
          },
          set: function (value) {
              this._setOption('autoHidePointMarkers', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "barGroupPadding", {
          /**
           * [descr:dxChartOptions.barGroupPadding]
          
           */
          get: function () {
              return this._getOption('barGroupPadding');
          },
          set: function (value) {
              this._setOption('barGroupPadding', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "barGroupWidth", {
          /**
           * [descr:dxChartOptions.barGroupWidth]
          
           */
          get: function () {
              return this._getOption('barGroupWidth');
          },
          set: function (value) {
              this._setOption('barGroupWidth', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "commonAnnotationSettings", {
          /**
           * [descr:dxChartOptions.commonAnnotationSettings]
          
           */
          get: function () {
              return this._getOption('commonAnnotationSettings');
          },
          set: function (value) {
              this._setOption('commonAnnotationSettings', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "commonAxisSettings", {
          /**
           * [descr:dxChartOptions.commonAxisSettings]
          
           */
          get: function () {
              return this._getOption('commonAxisSettings');
          },
          set: function (value) {
              this._setOption('commonAxisSettings', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "commonPaneSettings", {
          /**
           * [descr:dxChartOptions.commonPaneSettings]
          
           */
          get: function () {
              return this._getOption('commonPaneSettings');
          },
          set: function (value) {
              this._setOption('commonPaneSettings', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "commonSeriesSettings", {
          /**
           * [descr:dxChartOptions.commonSeriesSettings]
          
           */
          get: function () {
              return this._getOption('commonSeriesSettings');
          },
          set: function (value) {
              this._setOption('commonSeriesSettings', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "containerBackgroundColor", {
          /**
           * [descr:dxChartOptions.containerBackgroundColor]
          
           */
          get: function () {
              return this._getOption('containerBackgroundColor');
          },
          set: function (value) {
              this._setOption('containerBackgroundColor', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "crosshair", {
          /**
           * [descr:dxChartOptions.crosshair]
          
           */
          get: function () {
              return this._getOption('crosshair');
          },
          set: function (value) {
              this._setOption('crosshair', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "customizeAnnotation", {
          /**
           * [descr:dxChartOptions.customizeAnnotation]
          
           */
          get: function () {
              return this._getOption('customizeAnnotation');
          },
          set: function (value) {
              this._setOption('customizeAnnotation', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "customizeLabel", {
          /**
           * [descr:BaseChartOptions.customizeLabel]
          
           */
          get: function () {
              return this._getOption('customizeLabel');
          },
          set: function (value) {
              this._setOption('customizeLabel', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "customizePoint", {
          /**
           * [descr:BaseChartOptions.customizePoint]
          
           */
          get: function () {
              return this._getOption('customizePoint');
          },
          set: function (value) {
              this._setOption('customizePoint', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "dataPrepareSettings", {
          /**
           * [descr:dxChartOptions.dataPrepareSettings]
          
           */
          get: function () {
              return this._getOption('dataPrepareSettings');
          },
          set: function (value) {
              this._setOption('dataPrepareSettings', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "dataSource", {
          /**
           * [descr:BaseChartOptions.dataSource]
          
           */
          get: function () {
              return this._getOption('dataSource');
          },
          set: function (value) {
              this._setOption('dataSource', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "defaultPane", {
          /**
           * [descr:dxChartOptions.defaultPane]
          
           */
          get: function () {
              return this._getOption('defaultPane');
          },
          set: function (value) {
              this._setOption('defaultPane', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "disabled", {
          /**
           * [descr:BaseWidgetOptions.disabled]
          
           */
          get: function () {
              return this._getOption('disabled');
          },
          set: function (value) {
              this._setOption('disabled', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "elementAttr", {
          /**
           * [descr:DOMComponentOptions.elementAttr]
          
           */
          get: function () {
              return this._getOption('elementAttr');
          },
          set: function (value) {
              this._setOption('elementAttr', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "export", {
          /**
           * [descr:BaseWidgetOptions.export]
          
           */
          get: function () {
              return this._getOption('export');
          },
          set: function (value) {
              this._setOption('export', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "legend", {
          /**
           * [descr:dxChartOptions.legend]
          
           */
          get: function () {
              return this._getOption('legend');
          },
          set: function (value) {
              this._setOption('legend', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "loadingIndicator", {
          /**
           * [descr:BaseWidgetOptions.loadingIndicator]
          
           */
          get: function () {
              return this._getOption('loadingIndicator');
          },
          set: function (value) {
              this._setOption('loadingIndicator', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "margin", {
          /**
           * [descr:BaseWidgetOptions.margin]
          
           */
          get: function () {
              return this._getOption('margin');
          },
          set: function (value) {
              this._setOption('margin', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "maxBubbleSize", {
          /**
           * [descr:dxChartOptions.maxBubbleSize]
          
           */
          get: function () {
              return this._getOption('maxBubbleSize');
          },
          set: function (value) {
              this._setOption('maxBubbleSize', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "minBubbleSize", {
          /**
           * [descr:dxChartOptions.minBubbleSize]
          
           */
          get: function () {
              return this._getOption('minBubbleSize');
          },
          set: function (value) {
              this._setOption('minBubbleSize', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "negativesAsZeroes", {
          /**
           * [descr:dxChartOptions.negativesAsZeroes]
          
           */
          get: function () {
              return this._getOption('negativesAsZeroes');
          },
          set: function (value) {
              this._setOption('negativesAsZeroes', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "palette", {
          /**
           * [descr:BaseChartOptions.palette]
          
           */
          get: function () {
              return this._getOption('palette');
          },
          set: function (value) {
              this._setOption('palette', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "paletteExtensionMode", {
          /**
           * [descr:BaseChartOptions.paletteExtensionMode]
          
           */
          get: function () {
              return this._getOption('paletteExtensionMode');
          },
          set: function (value) {
              this._setOption('paletteExtensionMode', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "panes", {
          /**
           * [descr:dxChartOptions.panes]
          
           */
          get: function () {
              return this._getOption('panes');
          },
          set: function (value) {
              this._setOption('panes', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "pathModified", {
          /**
           * [descr:BaseWidgetOptions.pathModified]
          
           */
          get: function () {
              return this._getOption('pathModified');
          },
          set: function (value) {
              this._setOption('pathModified', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "pointSelectionMode", {
          /**
           * [descr:BaseChartOptions.pointSelectionMode]
          
           */
          get: function () {
              return this._getOption('pointSelectionMode');
          },
          set: function (value) {
              this._setOption('pointSelectionMode', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "redrawOnResize", {
          /**
           * [descr:BaseWidgetOptions.redrawOnResize]
          
           */
          get: function () {
              return this._getOption('redrawOnResize');
          },
          set: function (value) {
              this._setOption('redrawOnResize', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "resizePanesOnZoom", {
          /**
           * [descr:dxChartOptions.resizePanesOnZoom]
          
           */
          get: function () {
              return this._getOption('resizePanesOnZoom');
          },
          set: function (value) {
              this._setOption('resizePanesOnZoom', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "resolveLabelOverlapping", {
          /**
           * [descr:dxChartOptions.resolveLabelOverlapping]
          
           */
          get: function () {
              return this._getOption('resolveLabelOverlapping');
          },
          set: function (value) {
              this._setOption('resolveLabelOverlapping', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "rotated", {
          /**
           * [descr:dxChartOptions.rotated]
          
           */
          get: function () {
              return this._getOption('rotated');
          },
          set: function (value) {
              this._setOption('rotated', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "rtlEnabled", {
          /**
           * [descr:BaseWidgetOptions.rtlEnabled]
          
           */
          get: function () {
              return this._getOption('rtlEnabled');
          },
          set: function (value) {
              this._setOption('rtlEnabled', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "scrollBar", {
          /**
           * [descr:dxChartOptions.scrollBar]
          
           */
          get: function () {
              return this._getOption('scrollBar');
          },
          set: function (value) {
              this._setOption('scrollBar', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "series", {
          /**
           * [descr:dxChartOptions.series]
          
           */
          get: function () {
              return this._getOption('series');
          },
          set: function (value) {
              this._setOption('series', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "seriesSelectionMode", {
          /**
           * [descr:dxChartOptions.seriesSelectionMode]
          
           */
          get: function () {
              return this._getOption('seriesSelectionMode');
          },
          set: function (value) {
              this._setOption('seriesSelectionMode', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "seriesTemplate", {
          /**
           * [descr:dxChartOptions.seriesTemplate]
          
           */
          get: function () {
              return this._getOption('seriesTemplate');
          },
          set: function (value) {
              this._setOption('seriesTemplate', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "size", {
          /**
           * [descr:BaseWidgetOptions.size]
          
           */
          get: function () {
              return this._getOption('size');
          },
          set: function (value) {
              this._setOption('size', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "stickyHovering", {
          /**
           * [descr:dxChartOptions.stickyHovering]
          
           */
          get: function () {
              return this._getOption('stickyHovering');
          },
          set: function (value) {
              this._setOption('stickyHovering', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "synchronizeMultiAxes", {
          /**
           * [descr:dxChartOptions.synchronizeMultiAxes]
          
           */
          get: function () {
              return this._getOption('synchronizeMultiAxes');
          },
          set: function (value) {
              this._setOption('synchronizeMultiAxes', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "theme", {
          /**
           * [descr:BaseWidgetOptions.theme]
          
           */
          get: function () {
              return this._getOption('theme');
          },
          set: function (value) {
              this._setOption('theme', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "title", {
          /**
           * [descr:BaseWidgetOptions.title]
          
           */
          get: function () {
              return this._getOption('title');
          },
          set: function (value) {
              this._setOption('title', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "tooltip", {
          /**
           * [descr:dxChartOptions.tooltip]
          
           */
          get: function () {
              return this._getOption('tooltip');
          },
          set: function (value) {
              this._setOption('tooltip', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "valueAxis", {
          /**
           * [descr:dxChartOptions.valueAxis]
          
           */
          get: function () {
              return this._getOption('valueAxis');
          },
          set: function (value) {
              this._setOption('valueAxis', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "zoomAndPan", {
          /**
           * [descr:dxChartOptions.zoomAndPan]
          
           */
          get: function () {
              return this._getOption('zoomAndPan');
          },
          set: function (value) {
              this._setOption('zoomAndPan', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "annotationsChildren", {
          get: function () {
              return this._getOption('annotations');
          },
          set: function (value) {
              this.setChildren('annotations', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "panesChildren", {
          get: function () {
              return this._getOption('panes');
          },
          set: function (value) {
              this.setChildren('panes', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "seriesChildren", {
          get: function () {
              return this._getOption('series');
          },
          set: function (value) {
              this.setChildren('series', value);
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(DxChartComponent.prototype, "valueAxisChildren", {
          get: function () {
              return this._getOption('valueAxis');
          },
          set: function (value) {
              this.setChildren('valueAxis', value);
          },
          enumerable: false,
          configurable: true
      });
      DxChartComponent.prototype._createInstance = function (element, options) {
          return new DxChart__default["default"](element, options);
      };
      DxChartComponent.prototype.ngOnDestroy = function () {
          this._destroyWidget();
      };
      DxChartComponent.prototype.ngOnChanges = function (changes) {
          _super.prototype.ngOnChanges.call(this, changes);
          this.setupChanges('annotations', changes);
          this.setupChanges('dataSource', changes);
          this.setupChanges('palette', changes);
          this.setupChanges('panes', changes);
          this.setupChanges('series', changes);
          this.setupChanges('valueAxis', changes);
      };
      DxChartComponent.prototype.setupChanges = function (prop, changes) {
          if (!(prop in this._optionsToUpdate)) {
              this._idh.setup(prop, changes);
          }
      };
      DxChartComponent.prototype.ngDoCheck = function () {
          this._idh.doCheck('annotations');
          this._idh.doCheck('dataSource');
          this._idh.doCheck('palette');
          this._idh.doCheck('panes');
          this._idh.doCheck('series');
          this._idh.doCheck('valueAxis');
          this._watcherHelper.checkWatchers();
          _super.prototype.ngDoCheck.call(this);
          _super.prototype.clearChangedOptions.call(this);
      };
      DxChartComponent.prototype._setOption = function (name, value) {
          var isSetup = this._idh.setupSingle(name, value);
          var isChanged = this._idh.getChanges(name, value) !== null;
          if (isSetup || isChanged) {
              _super.prototype._setOption.call(this, name, value);
          }
      };
      return DxChartComponent;
  }(i1.DxComponent));
  /** @nocollapse */ DxChartComponent.fac = function DxChartComponent_Factory(t) { return new (t || DxChartComponent)(i0__namespace.directiveInject(i0__namespace.ElementRef), i0__namespace.directiveInject(i0__namespace.NgZone), i0__namespace.directiveInject(i1__namespace.DxTemplateHost), i0__namespace.directiveInject(i1__namespace.WatcherHelper), i0__namespace.directiveInject(i1__namespace.IterableDifferHelper), i0__namespace.directiveInject(i1__namespace.NestedOptionHost), i0__namespace.directiveInject(i2__namespace.TransferState), i0__namespace.directiveInject(i0.PLATFORM_ID)); };
  /** @nocollapse */ DxChartComponent.cmp = /** @pureOrBreakMyCode */ i0__namespace.defineComponent({ type: DxChartComponent, selectors: [["dx-chart"]], contentQueries: function DxChartComponent_ContentQueries(rf, ctx, dirIndex) {
          if (rf & 1) {
              i0__namespace.contentQuery(dirIndex, nested.DxiAnnotationComponent, 4);
              i0__namespace.contentQuery(dirIndex, nested.DxiPaneComponent, 4);
              i0__namespace.contentQuery(dirIndex, nested.DxiSeriesComponent, 4);
              i0__namespace.contentQuery(dirIndex, nested.DxiValueAxisComponent, 4);
          }
          if (rf & 2) {
              var _t = void 0;
              i0__namespace.queryRefresh(_t = i0__namespace.loadQuery()) && (ctx.annotationsChildren = _t);
              i0__namespace.queryRefresh(_t = i0__namespace.loadQuery()) && (ctx.panesChildren = _t);
              i0__namespace.queryRefresh(_t = i0__namespace.loadQuery()) && (ctx.seriesChildren = _t);
              i0__namespace.queryRefresh(_t = i0__namespace.loadQuery()) && (ctx.valueAxisChildren = _t);
          }
      }, inputs: { adaptiveLayout: "adaptiveLayout", adjustOnZoom: "adjustOnZoom", animation: "animation", annotations: "annotations", argumentAxis: "argumentAxis", autoHidePointMarkers: "autoHidePointMarkers", barGroupPadding: "barGroupPadding", barGroupWidth: "barGroupWidth", commonAnnotationSettings: "commonAnnotationSettings", commonAxisSettings: "commonAxisSettings", commonPaneSettings: "commonPaneSettings", commonSeriesSettings: "commonSeriesSettings", containerBackgroundColor: "containerBackgroundColor", crosshair: "crosshair", customizeAnnotation: "customizeAnnotation", customizeLabel: "customizeLabel", customizePoint: "customizePoint", dataPrepareSettings: "dataPrepareSettings", dataSource: "dataSource", defaultPane: "defaultPane", disabled: "disabled", elementAttr: "elementAttr", export: "export", legend: "legend", loadingIndicator: "loadingIndicator", margin: "margin", maxBubbleSize: "maxBubbleSize", minBubbleSize: "minBubbleSize", negativesAsZeroes: "negativesAsZeroes", palette: "palette", paletteExtensionMode: "paletteExtensionMode", panes: "panes", pathModified: "pathModified", pointSelectionMode: "pointSelectionMode", redrawOnResize: "redrawOnResize", resizePanesOnZoom: "resizePanesOnZoom", resolveLabelOverlapping: "resolveLabelOverlapping", rotated: "rotated", rtlEnabled: "rtlEnabled", scrollBar: "scrollBar", series: "series", seriesSelectionMode: "seriesSelectionMode", seriesTemplate: "seriesTemplate", size: "size", stickyHovering: "stickyHovering", synchronizeMultiAxes: "synchronizeMultiAxes", theme: "theme", title: "title", tooltip: "tooltip", valueAxis: "valueAxis", zoomAndPan: "zoomAndPan" }, outputs: { onArgumentAxisClick: "onArgumentAxisClick", onDisposing: "onDisposing", onDone: "onDone", onDrawn: "onDrawn", onExported: "onExported", onExporting: "onExporting", onFileSaving: "onFileSaving", onIncidentOccurred: "onIncidentOccurred", onInitialized: "onInitialized", onLegendClick: "onLegendClick", onOptionChanged: "onOptionChanged", onPointClick: "onPointClick", onPointHoverChanged: "onPointHoverChanged", onPointSelectionChanged: "onPointSelectionChanged", onSeriesClick: "onSeriesClick", onSeriesHoverChanged: "onSeriesHoverChanged", onSeriesSelectionChanged: "onSeriesSelectionChanged", onTooltipHidden: "onTooltipHidden", onTooltipShown: "onTooltipShown", onZoomEnd: "onZoomEnd", onZoomStart: "onZoomStart", adaptiveLayoutChange: "adaptiveLayoutChange", adjustOnZoomChange: "adjustOnZoomChange", animationChange: "animationChange", annotationsChange: "annotationsChange", argumentAxisChange: "argumentAxisChange", autoHidePointMarkersChange: "autoHidePointMarkersChange", barGroupPaddingChange: "barGroupPaddingChange", barGroupWidthChange: "barGroupWidthChange", commonAnnotationSettingsChange: "commonAnnotationSettingsChange", commonAxisSettingsChange: "commonAxisSettingsChange", commonPaneSettingsChange: "commonPaneSettingsChange", commonSeriesSettingsChange: "commonSeriesSettingsChange", containerBackgroundColorChange: "containerBackgroundColorChange", crosshairChange: "crosshairChange", customizeAnnotationChange: "customizeAnnotationChange", customizeLabelChange: "customizeLabelChange", customizePointChange: "customizePointChange", dataPrepareSettingsChange: "dataPrepareSettingsChange", dataSourceChange: "dataSourceChange", defaultPaneChange: "defaultPaneChange", disabledChange: "disabledChange", elementAttrChange: "elementAttrChange", exportChange: "exportChange", legendChange: "legendChange", loadingIndicatorChange: "loadingIndicatorChange", marginChange: "marginChange", maxBubbleSizeChange: "maxBubbleSizeChange", minBubbleSizeChange: "minBubbleSizeChange", negativesAsZeroesChange: "negativesAsZeroesChange", paletteChange: "paletteChange", paletteExtensionModeChange: "paletteExtensionModeChange", panesChange: "panesChange", pathModifiedChange: "pathModifiedChange", pointSelectionModeChange: "pointSelectionModeChange", redrawOnResizeChange: "redrawOnResizeChange", resizePanesOnZoomChange: "resizePanesOnZoomChange", resolveLabelOverlappingChange: "resolveLabelOverlappingChange", rotatedChange: "rotatedChange", rtlEnabledChange: "rtlEnabledChange", scrollBarChange: "scrollBarChange", seriesChange: "seriesChange", seriesSelectionModeChange: "seriesSelectionModeChange", seriesTemplateChange: "seriesTemplateChange", sizeChange: "sizeChange", stickyHoveringChange: "stickyHoveringChange", synchronizeMultiAxesChange: "synchronizeMultiAxesChange", themeChange: "themeChange", titleChange: "titleChange", tooltipChange: "tooltipChange", valueAxisChange: "valueAxisChange", zoomAndPanChange: "zoomAndPanChange" }, features: [i0__namespace.ProvidersFeature([
              i1.DxTemplateHost,
              i1.WatcherHelper,
              i1.NestedOptionHost,
              i1.IterableDifferHelper
          ]), i0__namespace.InheritDefinitionFeature, i0__namespace.NgOnChangesFeature], decls: 0, vars: 0, template: function DxChartComponent_Template(rf, ctx) { }, styles: ["[_nghost-%COMP%] {  display: block; }"] });
  (function () {
      (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.setClassMetadata(DxChartComponent, [{
              type: i0.Component,
              args: [{
                      selector: 'dx-chart',
                      template: '',
                      styles: [' :host {  display: block; }'],
                      providers: [
                          i1.DxTemplateHost,
                          i1.WatcherHelper,
                          i1.NestedOptionHost,
                          i1.IterableDifferHelper
                      ]
                  }]
          }], function () {
          return [{ type: i0__namespace.ElementRef }, { type: i0__namespace.NgZone }, { type: i1__namespace.DxTemplateHost }, { type: i1__namespace.WatcherHelper }, { type: i1__namespace.IterableDifferHelper }, { type: i1__namespace.NestedOptionHost }, { type: i2__namespace.TransferState }, { type: undefined, decorators: [{
                          type: i0.Inject,
                          args: [i0.PLATFORM_ID]
                      }] }];
      }, { adaptiveLayout: [{
                  type: i0.Input
              }], adjustOnZoom: [{
                  type: i0.Input
              }], animation: [{
                  type: i0.Input
              }], annotations: [{
                  type: i0.Input
              }], argumentAxis: [{
                  type: i0.Input
              }], autoHidePointMarkers: [{
                  type: i0.Input
              }], barGroupPadding: [{
                  type: i0.Input
              }], barGroupWidth: [{
                  type: i0.Input
              }], commonAnnotationSettings: [{
                  type: i0.Input
              }], commonAxisSettings: [{
                  type: i0.Input
              }], commonPaneSettings: [{
                  type: i0.Input
              }], commonSeriesSettings: [{
                  type: i0.Input
              }], containerBackgroundColor: [{
                  type: i0.Input
              }], crosshair: [{
                  type: i0.Input
              }], customizeAnnotation: [{
                  type: i0.Input
              }], customizeLabel: [{
                  type: i0.Input
              }], customizePoint: [{
                  type: i0.Input
              }], dataPrepareSettings: [{
                  type: i0.Input
              }], dataSource: [{
                  type: i0.Input
              }], defaultPane: [{
                  type: i0.Input
              }], disabled: [{
                  type: i0.Input
              }], elementAttr: [{
                  type: i0.Input
              }], export: [{
                  type: i0.Input
              }], legend: [{
                  type: i0.Input
              }], loadingIndicator: [{
                  type: i0.Input
              }], margin: [{
                  type: i0.Input
              }], maxBubbleSize: [{
                  type: i0.Input
              }], minBubbleSize: [{
                  type: i0.Input
              }], negativesAsZeroes: [{
                  type: i0.Input
              }], palette: [{
                  type: i0.Input
              }], paletteExtensionMode: [{
                  type: i0.Input
              }], panes: [{
                  type: i0.Input
              }], pathModified: [{
                  type: i0.Input
              }], pointSelectionMode: [{
                  type: i0.Input
              }], redrawOnResize: [{
                  type: i0.Input
              }], resizePanesOnZoom: [{
                  type: i0.Input
              }], resolveLabelOverlapping: [{
                  type: i0.Input
              }], rotated: [{
                  type: i0.Input
              }], rtlEnabled: [{
                  type: i0.Input
              }], scrollBar: [{
                  type: i0.Input
              }], series: [{
                  type: i0.Input
              }], seriesSelectionMode: [{
                  type: i0.Input
              }], seriesTemplate: [{
                  type: i0.Input
              }], size: [{
                  type: i0.Input
              }], stickyHovering: [{
                  type: i0.Input
              }], synchronizeMultiAxes: [{
                  type: i0.Input
              }], theme: [{
                  type: i0.Input
              }], title: [{
                  type: i0.Input
              }], tooltip: [{
                  type: i0.Input
              }], valueAxis: [{
                  type: i0.Input
              }], zoomAndPan: [{
                  type: i0.Input
              }], onArgumentAxisClick: [{
                  type: i0.Output
              }], onDisposing: [{
                  type: i0.Output
              }], onDone: [{
                  type: i0.Output
              }], onDrawn: [{
                  type: i0.Output
              }], onExported: [{
                  type: i0.Output
              }], onExporting: [{
                  type: i0.Output
              }], onFileSaving: [{
                  type: i0.Output
              }], onIncidentOccurred: [{
                  type: i0.Output
              }], onInitialized: [{
                  type: i0.Output
              }], onLegendClick: [{
                  type: i0.Output
              }], onOptionChanged: [{
                  type: i0.Output
              }], onPointClick: [{
                  type: i0.Output
              }], onPointHoverChanged: [{
                  type: i0.Output
              }], onPointSelectionChanged: [{
                  type: i0.Output
              }], onSeriesClick: [{
                  type: i0.Output
              }], onSeriesHoverChanged: [{
                  type: i0.Output
              }], onSeriesSelectionChanged: [{
                  type: i0.Output
              }], onTooltipHidden: [{
                  type: i0.Output
              }], onTooltipShown: [{
                  type: i0.Output
              }], onZoomEnd: [{
                  type: i0.Output
              }], onZoomStart: [{
                  type: i0.Output
              }], adaptiveLayoutChange: [{
                  type: i0.Output
              }], adjustOnZoomChange: [{
                  type: i0.Output
              }], animationChange: [{
                  type: i0.Output
              }], annotationsChange: [{
                  type: i0.Output
              }], argumentAxisChange: [{
                  type: i0.Output
              }], autoHidePointMarkersChange: [{
                  type: i0.Output
              }], barGroupPaddingChange: [{
                  type: i0.Output
              }], barGroupWidthChange: [{
                  type: i0.Output
              }], commonAnnotationSettingsChange: [{
                  type: i0.Output
              }], commonAxisSettingsChange: [{
                  type: i0.Output
              }], commonPaneSettingsChange: [{
                  type: i0.Output
              }], commonSeriesSettingsChange: [{
                  type: i0.Output
              }], containerBackgroundColorChange: [{
                  type: i0.Output
              }], crosshairChange: [{
                  type: i0.Output
              }], customizeAnnotationChange: [{
                  type: i0.Output
              }], customizeLabelChange: [{
                  type: i0.Output
              }], customizePointChange: [{
                  type: i0.Output
              }], dataPrepareSettingsChange: [{
                  type: i0.Output
              }], dataSourceChange: [{
                  type: i0.Output
              }], defaultPaneChange: [{
                  type: i0.Output
              }], disabledChange: [{
                  type: i0.Output
              }], elementAttrChange: [{
                  type: i0.Output
              }], exportChange: [{
                  type: i0.Output
              }], legendChange: [{
                  type: i0.Output
              }], loadingIndicatorChange: [{
                  type: i0.Output
              }], marginChange: [{
                  type: i0.Output
              }], maxBubbleSizeChange: [{
                  type: i0.Output
              }], minBubbleSizeChange: [{
                  type: i0.Output
              }], negativesAsZeroesChange: [{
                  type: i0.Output
              }], paletteChange: [{
                  type: i0.Output
              }], paletteExtensionModeChange: [{
                  type: i0.Output
              }], panesChange: [{
                  type: i0.Output
              }], pathModifiedChange: [{
                  type: i0.Output
              }], pointSelectionModeChange: [{
                  type: i0.Output
              }], redrawOnResizeChange: [{
                  type: i0.Output
              }], resizePanesOnZoomChange: [{
                  type: i0.Output
              }], resolveLabelOverlappingChange: [{
                  type: i0.Output
              }], rotatedChange: [{
                  type: i0.Output
              }], rtlEnabledChange: [{
                  type: i0.Output
              }], scrollBarChange: [{
                  type: i0.Output
              }], seriesChange: [{
                  type: i0.Output
              }], seriesSelectionModeChange: [{
                  type: i0.Output
              }], seriesTemplateChange: [{
                  type: i0.Output
              }], sizeChange: [{
                  type: i0.Output
              }], stickyHoveringChange: [{
                  type: i0.Output
              }], synchronizeMultiAxesChange: [{
                  type: i0.Output
              }], themeChange: [{
                  type: i0.Output
              }], titleChange: [{
                  type: i0.Output
              }], tooltipChange: [{
                  type: i0.Output
              }], valueAxisChange: [{
                  type: i0.Output
              }], zoomAndPanChange: [{
                  type: i0.Output
              }], annotationsChildren: [{
                  type: i0.ContentChildren,
                  args: [nested.DxiAnnotationComponent]
              }], panesChildren: [{
                  type: i0.ContentChildren,
                  args: [nested.DxiPaneComponent]
              }], seriesChildren: [{
                  type: i0.ContentChildren,
                  args: [nested.DxiSeriesComponent]
              }], valueAxisChildren: [{
                  type: i0.ContentChildren,
                  args: [nested.DxiValueAxisComponent]
              }] });
  })();
  var DxChartModule = /** @class */ (function () {
      function DxChartModule() {
      }
      return DxChartModule;
  }());
  /** @nocollapse */ DxChartModule.fac = function DxChartModule_Factory(t) { return new (t || DxChartModule)(); };
  /** @nocollapse */ DxChartModule.mod = /** @pureOrBreakMyCode */ i0__namespace.defineNgModule({ type: DxChartModule });
  /** @nocollapse */ DxChartModule.inj = /** @pureOrBreakMyCode */ i0__namespace.defineInjector({ imports: [[
              nested.DxoAdaptiveLayoutModule,
              nested.DxoAnimationModule,
              nested.DxiAnnotationModule,
              nested.DxoBorderModule,
              nested.DxoFontModule,
              nested.DxoImageModule,
              nested.DxoShadowModule,
              nested.DxoArgumentAxisModule,
              nested.DxoAggregationIntervalModule,
              nested.DxiBreakModule,
              nested.DxoBreakStyleModule,
              nested.DxiConstantLineModule,
              nested.DxoLabelModule,
              nested.DxoConstantLineStyleModule,
              nested.DxoGridModule,
              nested.DxoFormatModule,
              nested.DxoMinorGridModule,
              nested.DxoMinorTickModule,
              nested.DxoMinorTickIntervalModule,
              nested.DxoMinVisualRangeLengthModule,
              nested.DxiStripModule,
              nested.DxoStripStyleModule,
              nested.DxoTickModule,
              nested.DxoTickIntervalModule,
              nested.DxoTitleModule,
              nested.DxoCommonAnnotationSettingsModule,
              nested.DxoCommonAxisSettingsModule,
              nested.DxoCommonPaneSettingsModule,
              nested.DxoBackgroundColorModule,
              nested.DxoCommonSeriesSettingsModule,
              nested.DxoAggregationModule,
              nested.DxoAreaModule,
              nested.DxoHoverStyleModule,
              nested.DxoHatchingModule,
              nested.DxoConnectorModule,
              nested.DxoPointModule,
              nested.DxoHeightModule,
              nested.DxoUrlModule,
              nested.DxoWidthModule,
              nested.DxoSelectionStyleModule,
              nested.DxoReductionModule,
              nested.DxoValueErrorBarModule,
              nested.DxoBarModule,
              nested.DxoBubbleModule,
              nested.DxoCandlestickModule,
              nested.DxoColorModule,
              nested.DxoFullstackedareaModule,
              nested.DxoFullstackedbarModule,
              nested.DxoFullstackedlineModule,
              nested.DxoFullstackedsplineModule,
              nested.DxoFullstackedsplineareaModule,
              nested.DxoArgumentFormatModule,
              nested.DxoLineModule,
              nested.DxoRangeareaModule,
              nested.DxoRangebarModule,
              nested.DxoScatterModule,
              nested.DxoSplineModule,
              nested.DxoSplineareaModule,
              nested.DxoStackedareaModule,
              nested.DxoStackedbarModule,
              nested.DxoStackedlineModule,
              nested.DxoStackedsplineModule,
              nested.DxoStackedsplineareaModule,
              nested.DxoStepareaModule,
              nested.DxoSteplineModule,
              nested.DxoStockModule,
              nested.DxoCrosshairModule,
              nested.DxoHorizontalLineModule,
              nested.DxoVerticalLineModule,
              nested.DxoDataPrepareSettingsModule,
              nested.DxoExportModule,
              nested.DxoLegendModule,
              nested.DxoMarginModule,
              nested.DxoSubtitleModule,
              nested.DxoLoadingIndicatorModule,
              nested.DxiPaneModule,
              nested.DxoScrollBarModule,
              nested.DxiSeriesModule,
              nested.DxoSeriesTemplateModule,
              nested.DxoSizeModule,
              nested.DxoTooltipModule,
              nested.DxiValueAxisModule,
              nested.DxoZoomAndPanModule,
              nested.DxoDragBoxStyleModule,
              i1.DxIntegrationModule,
              i1.DxTemplateModule
          ], nested.DxoAdaptiveLayoutModule,
          nested.DxoAnimationModule,
          nested.DxiAnnotationModule,
          nested.DxoBorderModule,
          nested.DxoFontModule,
          nested.DxoImageModule,
          nested.DxoShadowModule,
          nested.DxoArgumentAxisModule,
          nested.DxoAggregationIntervalModule,
          nested.DxiBreakModule,
          nested.DxoBreakStyleModule,
          nested.DxiConstantLineModule,
          nested.DxoLabelModule,
          nested.DxoConstantLineStyleModule,
          nested.DxoGridModule,
          nested.DxoFormatModule,
          nested.DxoMinorGridModule,
          nested.DxoMinorTickModule,
          nested.DxoMinorTickIntervalModule,
          nested.DxoMinVisualRangeLengthModule,
          nested.DxiStripModule,
          nested.DxoStripStyleModule,
          nested.DxoTickModule,
          nested.DxoTickIntervalModule,
          nested.DxoTitleModule,
          nested.DxoCommonAnnotationSettingsModule,
          nested.DxoCommonAxisSettingsModule,
          nested.DxoCommonPaneSettingsModule,
          nested.DxoBackgroundColorModule,
          nested.DxoCommonSeriesSettingsModule,
          nested.DxoAggregationModule,
          nested.DxoAreaModule,
          nested.DxoHoverStyleModule,
          nested.DxoHatchingModule,
          nested.DxoConnectorModule,
          nested.DxoPointModule,
          nested.DxoHeightModule,
          nested.DxoUrlModule,
          nested.DxoWidthModule,
          nested.DxoSelectionStyleModule,
          nested.DxoReductionModule,
          nested.DxoValueErrorBarModule,
          nested.DxoBarModule,
          nested.DxoBubbleModule,
          nested.DxoCandlestickModule,
          nested.DxoColorModule,
          nested.DxoFullstackedareaModule,
          nested.DxoFullstackedbarModule,
          nested.DxoFullstackedlineModule,
          nested.DxoFullstackedsplineModule,
          nested.DxoFullstackedsplineareaModule,
          nested.DxoArgumentFormatModule,
          nested.DxoLineModule,
          nested.DxoRangeareaModule,
          nested.DxoRangebarModule,
          nested.DxoScatterModule,
          nested.DxoSplineModule,
          nested.DxoSplineareaModule,
          nested.DxoStackedareaModule,
          nested.DxoStackedbarModule,
          nested.DxoStackedlineModule,
          nested.DxoStackedsplineModule,
          nested.DxoStackedsplineareaModule,
          nested.DxoStepareaModule,
          nested.DxoSteplineModule,
          nested.DxoStockModule,
          nested.DxoCrosshairModule,
          nested.DxoHorizontalLineModule,
          nested.DxoVerticalLineModule,
          nested.DxoDataPrepareSettingsModule,
          nested.DxoExportModule,
          nested.DxoLegendModule,
          nested.DxoMarginModule,
          nested.DxoSubtitleModule,
          nested.DxoLoadingIndicatorModule,
          nested.DxiPaneModule,
          nested.DxoScrollBarModule,
          nested.DxiSeriesModule,
          nested.DxoSeriesTemplateModule,
          nested.DxoSizeModule,
          nested.DxoTooltipModule,
          nested.DxiValueAxisModule,
          nested.DxoZoomAndPanModule,
          nested.DxoDragBoxStyleModule,
          i1.DxTemplateModule] });
  (function () {
      (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.setClassMetadata(DxChartModule, [{
              type: i0.NgModule,
              args: [{
                      imports: [
                          nested.DxoAdaptiveLayoutModule,
                          nested.DxoAnimationModule,
                          nested.DxiAnnotationModule,
                          nested.DxoBorderModule,
                          nested.DxoFontModule,
                          nested.DxoImageModule,
                          nested.DxoShadowModule,
                          nested.DxoArgumentAxisModule,
                          nested.DxoAggregationIntervalModule,
                          nested.DxiBreakModule,
                          nested.DxoBreakStyleModule,
                          nested.DxiConstantLineModule,
                          nested.DxoLabelModule,
                          nested.DxoConstantLineStyleModule,
                          nested.DxoGridModule,
                          nested.DxoFormatModule,
                          nested.DxoMinorGridModule,
                          nested.DxoMinorTickModule,
                          nested.DxoMinorTickIntervalModule,
                          nested.DxoMinVisualRangeLengthModule,
                          nested.DxiStripModule,
                          nested.DxoStripStyleModule,
                          nested.DxoTickModule,
                          nested.DxoTickIntervalModule,
                          nested.DxoTitleModule,
                          nested.DxoCommonAnnotationSettingsModule,
                          nested.DxoCommonAxisSettingsModule,
                          nested.DxoCommonPaneSettingsModule,
                          nested.DxoBackgroundColorModule,
                          nested.DxoCommonSeriesSettingsModule,
                          nested.DxoAggregationModule,
                          nested.DxoAreaModule,
                          nested.DxoHoverStyleModule,
                          nested.DxoHatchingModule,
                          nested.DxoConnectorModule,
                          nested.DxoPointModule,
                          nested.DxoHeightModule,
                          nested.DxoUrlModule,
                          nested.DxoWidthModule,
                          nested.DxoSelectionStyleModule,
                          nested.DxoReductionModule,
                          nested.DxoValueErrorBarModule,
                          nested.DxoBarModule,
                          nested.DxoBubbleModule,
                          nested.DxoCandlestickModule,
                          nested.DxoColorModule,
                          nested.DxoFullstackedareaModule,
                          nested.DxoFullstackedbarModule,
                          nested.DxoFullstackedlineModule,
                          nested.DxoFullstackedsplineModule,
                          nested.DxoFullstackedsplineareaModule,
                          nested.DxoArgumentFormatModule,
                          nested.DxoLineModule,
                          nested.DxoRangeareaModule,
                          nested.DxoRangebarModule,
                          nested.DxoScatterModule,
                          nested.DxoSplineModule,
                          nested.DxoSplineareaModule,
                          nested.DxoStackedareaModule,
                          nested.DxoStackedbarModule,
                          nested.DxoStackedlineModule,
                          nested.DxoStackedsplineModule,
                          nested.DxoStackedsplineareaModule,
                          nested.DxoStepareaModule,
                          nested.DxoSteplineModule,
                          nested.DxoStockModule,
                          nested.DxoCrosshairModule,
                          nested.DxoHorizontalLineModule,
                          nested.DxoVerticalLineModule,
                          nested.DxoDataPrepareSettingsModule,
                          nested.DxoExportModule,
                          nested.DxoLegendModule,
                          nested.DxoMarginModule,
                          nested.DxoSubtitleModule,
                          nested.DxoLoadingIndicatorModule,
                          nested.DxiPaneModule,
                          nested.DxoScrollBarModule,
                          nested.DxiSeriesModule,
                          nested.DxoSeriesTemplateModule,
                          nested.DxoSizeModule,
                          nested.DxoTooltipModule,
                          nested.DxiValueAxisModule,
                          nested.DxoZoomAndPanModule,
                          nested.DxoDragBoxStyleModule,
                          i1.DxIntegrationModule,
                          i1.DxTemplateModule
                      ],
                      declarations: [
                          DxChartComponent
                      ],
                      exports: [
                          DxChartComponent,
                          nested.DxoAdaptiveLayoutModule,
                          nested.DxoAnimationModule,
                          nested.DxiAnnotationModule,
                          nested.DxoBorderModule,
                          nested.DxoFontModule,
                          nested.DxoImageModule,
                          nested.DxoShadowModule,
                          nested.DxoArgumentAxisModule,
                          nested.DxoAggregationIntervalModule,
                          nested.DxiBreakModule,
                          nested.DxoBreakStyleModule,
                          nested.DxiConstantLineModule,
                          nested.DxoLabelModule,
                          nested.DxoConstantLineStyleModule,
                          nested.DxoGridModule,
                          nested.DxoFormatModule,
                          nested.DxoMinorGridModule,
                          nested.DxoMinorTickModule,
                          nested.DxoMinorTickIntervalModule,
                          nested.DxoMinVisualRangeLengthModule,
                          nested.DxiStripModule,
                          nested.DxoStripStyleModule,
                          nested.DxoTickModule,
                          nested.DxoTickIntervalModule,
                          nested.DxoTitleModule,
                          nested.DxoCommonAnnotationSettingsModule,
                          nested.DxoCommonAxisSettingsModule,
                          nested.DxoCommonPaneSettingsModule,
                          nested.DxoBackgroundColorModule,
                          nested.DxoCommonSeriesSettingsModule,
                          nested.DxoAggregationModule,
                          nested.DxoAreaModule,
                          nested.DxoHoverStyleModule,
                          nested.DxoHatchingModule,
                          nested.DxoConnectorModule,
                          nested.DxoPointModule,
                          nested.DxoHeightModule,
                          nested.DxoUrlModule,
                          nested.DxoWidthModule,
                          nested.DxoSelectionStyleModule,
                          nested.DxoReductionModule,
                          nested.DxoValueErrorBarModule,
                          nested.DxoBarModule,
                          nested.DxoBubbleModule,
                          nested.DxoCandlestickModule,
                          nested.DxoColorModule,
                          nested.DxoFullstackedareaModule,
                          nested.DxoFullstackedbarModule,
                          nested.DxoFullstackedlineModule,
                          nested.DxoFullstackedsplineModule,
                          nested.DxoFullstackedsplineareaModule,
                          nested.DxoArgumentFormatModule,
                          nested.DxoLineModule,
                          nested.DxoRangeareaModule,
                          nested.DxoRangebarModule,
                          nested.DxoScatterModule,
                          nested.DxoSplineModule,
                          nested.DxoSplineareaModule,
                          nested.DxoStackedareaModule,
                          nested.DxoStackedbarModule,
                          nested.DxoStackedlineModule,
                          nested.DxoStackedsplineModule,
                          nested.DxoStackedsplineareaModule,
                          nested.DxoStepareaModule,
                          nested.DxoSteplineModule,
                          nested.DxoStockModule,
                          nested.DxoCrosshairModule,
                          nested.DxoHorizontalLineModule,
                          nested.DxoVerticalLineModule,
                          nested.DxoDataPrepareSettingsModule,
                          nested.DxoExportModule,
                          nested.DxoLegendModule,
                          nested.DxoMarginModule,
                          nested.DxoSubtitleModule,
                          nested.DxoLoadingIndicatorModule,
                          nested.DxiPaneModule,
                          nested.DxoScrollBarModule,
                          nested.DxiSeriesModule,
                          nested.DxoSeriesTemplateModule,
                          nested.DxoSizeModule,
                          nested.DxoTooltipModule,
                          nested.DxiValueAxisModule,
                          nested.DxoZoomAndPanModule,
                          nested.DxoDragBoxStyleModule,
                          i1.DxTemplateModule
                      ]
                  }]
          }], null, null);
  })();
  (function () {
      (typeof ngJitMode === "undefined" || ngJitMode) && i0__namespace.setNgModuleScope(DxChartModule, { declarations: [DxChartComponent], imports: [nested.DxoAdaptiveLayoutModule,
              nested.DxoAnimationModule,
              nested.DxiAnnotationModule,
              nested.DxoBorderModule,
              nested.DxoFontModule,
              nested.DxoImageModule,
              nested.DxoShadowModule,
              nested.DxoArgumentAxisModule,
              nested.DxoAggregationIntervalModule,
              nested.DxiBreakModule,
              nested.DxoBreakStyleModule,
              nested.DxiConstantLineModule,
              nested.DxoLabelModule,
              nested.DxoConstantLineStyleModule,
              nested.DxoGridModule,
              nested.DxoFormatModule,
              nested.DxoMinorGridModule,
              nested.DxoMinorTickModule,
              nested.DxoMinorTickIntervalModule,
              nested.DxoMinVisualRangeLengthModule,
              nested.DxiStripModule,
              nested.DxoStripStyleModule,
              nested.DxoTickModule,
              nested.DxoTickIntervalModule,
              nested.DxoTitleModule,
              nested.DxoCommonAnnotationSettingsModule,
              nested.DxoCommonAxisSettingsModule,
              nested.DxoCommonPaneSettingsModule,
              nested.DxoBackgroundColorModule,
              nested.DxoCommonSeriesSettingsModule,
              nested.DxoAggregationModule,
              nested.DxoAreaModule,
              nested.DxoHoverStyleModule,
              nested.DxoHatchingModule,
              nested.DxoConnectorModule,
              nested.DxoPointModule,
              nested.DxoHeightModule,
              nested.DxoUrlModule,
              nested.DxoWidthModule,
              nested.DxoSelectionStyleModule,
              nested.DxoReductionModule,
              nested.DxoValueErrorBarModule,
              nested.DxoBarModule,
              nested.DxoBubbleModule,
              nested.DxoCandlestickModule,
              nested.DxoColorModule,
              nested.DxoFullstackedareaModule,
              nested.DxoFullstackedbarModule,
              nested.DxoFullstackedlineModule,
              nested.DxoFullstackedsplineModule,
              nested.DxoFullstackedsplineareaModule,
              nested.DxoArgumentFormatModule,
              nested.DxoLineModule,
              nested.DxoRangeareaModule,
              nested.DxoRangebarModule,
              nested.DxoScatterModule,
              nested.DxoSplineModule,
              nested.DxoSplineareaModule,
              nested.DxoStackedareaModule,
              nested.DxoStackedbarModule,
              nested.DxoStackedlineModule,
              nested.DxoStackedsplineModule,
              nested.DxoStackedsplineareaModule,
              nested.DxoStepareaModule,
              nested.DxoSteplineModule,
              nested.DxoStockModule,
              nested.DxoCrosshairModule,
              nested.DxoHorizontalLineModule,
              nested.DxoVerticalLineModule,
              nested.DxoDataPrepareSettingsModule,
              nested.DxoExportModule,
              nested.DxoLegendModule,
              nested.DxoMarginModule,
              nested.DxoSubtitleModule,
              nested.DxoLoadingIndicatorModule,
              nested.DxiPaneModule,
              nested.DxoScrollBarModule,
              nested.DxiSeriesModule,
              nested.DxoSeriesTemplateModule,
              nested.DxoSizeModule,
              nested.DxoTooltipModule,
              nested.DxiValueAxisModule,
              nested.DxoZoomAndPanModule,
              nested.DxoDragBoxStyleModule,
              i1.DxIntegrationModule,
              i1.DxTemplateModule], exports: [DxChartComponent, nested.DxoAdaptiveLayoutModule,
              nested.DxoAnimationModule,
              nested.DxiAnnotationModule,
              nested.DxoBorderModule,
              nested.DxoFontModule,
              nested.DxoImageModule,
              nested.DxoShadowModule,
              nested.DxoArgumentAxisModule,
              nested.DxoAggregationIntervalModule,
              nested.DxiBreakModule,
              nested.DxoBreakStyleModule,
              nested.DxiConstantLineModule,
              nested.DxoLabelModule,
              nested.DxoConstantLineStyleModule,
              nested.DxoGridModule,
              nested.DxoFormatModule,
              nested.DxoMinorGridModule,
              nested.DxoMinorTickModule,
              nested.DxoMinorTickIntervalModule,
              nested.DxoMinVisualRangeLengthModule,
              nested.DxiStripModule,
              nested.DxoStripStyleModule,
              nested.DxoTickModule,
              nested.DxoTickIntervalModule,
              nested.DxoTitleModule,
              nested.DxoCommonAnnotationSettingsModule,
              nested.DxoCommonAxisSettingsModule,
              nested.DxoCommonPaneSettingsModule,
              nested.DxoBackgroundColorModule,
              nested.DxoCommonSeriesSettingsModule,
              nested.DxoAggregationModule,
              nested.DxoAreaModule,
              nested.DxoHoverStyleModule,
              nested.DxoHatchingModule,
              nested.DxoConnectorModule,
              nested.DxoPointModule,
              nested.DxoHeightModule,
              nested.DxoUrlModule,
              nested.DxoWidthModule,
              nested.DxoSelectionStyleModule,
              nested.DxoReductionModule,
              nested.DxoValueErrorBarModule,
              nested.DxoBarModule,
              nested.DxoBubbleModule,
              nested.DxoCandlestickModule,
              nested.DxoColorModule,
              nested.DxoFullstackedareaModule,
              nested.DxoFullstackedbarModule,
              nested.DxoFullstackedlineModule,
              nested.DxoFullstackedsplineModule,
              nested.DxoFullstackedsplineareaModule,
              nested.DxoArgumentFormatModule,
              nested.DxoLineModule,
              nested.DxoRangeareaModule,
              nested.DxoRangebarModule,
              nested.DxoScatterModule,
              nested.DxoSplineModule,
              nested.DxoSplineareaModule,
              nested.DxoStackedareaModule,
              nested.DxoStackedbarModule,
              nested.DxoStackedlineModule,
              nested.DxoStackedsplineModule,
              nested.DxoStackedsplineareaModule,
              nested.DxoStepareaModule,
              nested.DxoSteplineModule,
              nested.DxoStockModule,
              nested.DxoCrosshairModule,
              nested.DxoHorizontalLineModule,
              nested.DxoVerticalLineModule,
              nested.DxoDataPrepareSettingsModule,
              nested.DxoExportModule,
              nested.DxoLegendModule,
              nested.DxoMarginModule,
              nested.DxoSubtitleModule,
              nested.DxoLoadingIndicatorModule,
              nested.DxiPaneModule,
              nested.DxoScrollBarModule,
              nested.DxiSeriesModule,
              nested.DxoSeriesTemplateModule,
              nested.DxoSizeModule,
              nested.DxoTooltipModule,
              nested.DxiValueAxisModule,
              nested.DxoZoomAndPanModule,
              nested.DxoDragBoxStyleModule,
              i1.DxTemplateModule] });
  })();

  /**
   * Generated bundle index. Do not edit.
   */

  exports.DxChartComponent = DxChartComponent;
  exports.DxChartModule = DxChartModule;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
