// ngx-editor/esm2022/lib/commands/Heading.mjs
import { setBlockType } from 'prosemirror-commands';
import { getSelectionNodes } from 'ngx-editor/helpers';
class Heading {
    constructor(level) {
        this.level = level;
    }
    apply() {
        return (state, dispatch) => {
            const { schema } = state;
            const type = schema.nodes['heading'];
            if (!type) {
                return false;
            }
            return setBlockType(type)(state, dispatch);
        };
    }
    toggle() {
        return (state, dispatch) => {
            const { schema, selection, doc } = state;
            const type = schema.nodes['heading'];
            if (!type) {
                return false;
            }
            const nodePos = selection.$from.before(1);
            const node = doc.nodeAt(nodePos);
            const attrs = node?.attrs ?? {};
            if (this.isActive(state)) {
                return setBlockType(schema.nodes['paragraph'], attrs)(state, dispatch);
            }
            return setBlockType(type, { ...attrs, level: this.level })(state, dispatch);
        };
    }
    isActive(state) {
        const { schema } = state;
        const nodesInSelection = getSelectionNodes(state);
        const type = schema.nodes['heading'];
        if (!type) {
            return false;
        }
        const supportedNodes = [
            type,
            schema.nodes['text'],
            schema.nodes['blockquote'],
        ];
        // heading is a text node
        // don't mark as active when it has more nodes
        const nodes = nodesInSelection.filter((node) => {
            return supportedNodes.includes(node.type);
        });
        const acitveNode = nodes.find((node) => {
            return node.attrs['level'] === this.level;
        });
        return Boolean(acitveNode);
    }
    canExecute(state) {
        return this.toggle()(state);
    }
}
export default Heading;
